<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>AnnotateFeatures — AnnotateFeatures • scop</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="AnnotateFeatures — AnnotateFeatures"><meta name="description" content="Annotate features in a Seurat object with additional metadata from databases or a GTF file."><meta property="og:description" content="Annotate features in a Seurat object with additional metadata from databases or a GTF file."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scop</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/mengxu98/scop" title="GitHub" aria-label="GitHub repository"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mengxu98" title="MengXu" aria-label="MengXu's GitHub profile"><span class="fa fas fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>AnnotateFeatures</h1>
      <small class="dont-index">Source: <a href="https://github.com/mengxu98/scop/blob/main/R/AnnotateFeatures.R" class="external-link"><code>R/AnnotateFeatures.R</code></a></small>
      <div class="d-none name"><code>AnnotateFeatures.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Annotate features in a Seurat object with additional metadata from databases or a GTF file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">AnnotateFeatures</span><span class="op">(</span></span>
<span>  <span class="va">srt</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Homo_sapiens"</span>,</span>
<span>  IDtype <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"symbol"</span>, <span class="st">"ensembl_id"</span>, <span class="st">"entrez_id"</span><span class="op">)</span>,</span>
<span>  db <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  db_update <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  db_version <span class="op">=</span> <span class="st">"latest"</span>,</span>
<span>  convert_species <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  Ensembl_version <span class="op">=</span> <span class="fl">103</span>,</span>
<span>  mirror <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gtf <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  merge_gtf_by <span class="op">=</span> <span class="st">"gene_name"</span>,</span>
<span>  columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"seqname"</span>, <span class="st">"feature"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"strand"</span>, <span class="st">"gene_id"</span>, <span class="st">"gene_name"</span>,</span>
<span>    <span class="st">"gene_type"</span><span class="op">)</span>,</span>
<span>  assays <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  overwrite <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-srt">srt<a class="anchor" aria-label="anchor" href="#arg-srt"></a></dt>
<dd><p>Seurat object to be annotated.</p></dd>


<dt id="arg-species">species<a class="anchor" aria-label="anchor" href="#arg-species"></a></dt>
<dd><p>Name of the species to be used for annotation.
Default is "Homo_sapiens".</p></dd>


<dt id="arg-idtype">IDtype<a class="anchor" aria-label="anchor" href="#arg-idtype"></a></dt>
<dd><p>Type of identifier to use for annotation.
Default is "symbol" with options "symbol", "ensembl_id", and "entrez_id".</p></dd>


<dt id="arg-db">db<a class="anchor" aria-label="anchor" href="#arg-db"></a></dt>
<dd><p>Vector of database names to be used for annotation.
Default is NULL.</p></dd>


<dt id="arg-db-update">db_update<a class="anchor" aria-label="anchor" href="#arg-db-update"></a></dt>
<dd><p>Logical value indicating whether to update the database.
Default is FALSE.</p></dd>


<dt id="arg-db-version">db_version<a class="anchor" aria-label="anchor" href="#arg-db-version"></a></dt>
<dd><p>Version of the database to use.
Default is "latest".</p></dd>


<dt id="arg-convert-species">convert_species<a class="anchor" aria-label="anchor" href="#arg-convert-species"></a></dt>
<dd><p>A logical value indicating whether to use a species-converted database when the annotation is missing for the specified species.
The default value is TRUE.</p></dd>


<dt id="arg-ensembl-version">Ensembl_version<a class="anchor" aria-label="anchor" href="#arg-ensembl-version"></a></dt>
<dd><p>Version of the Ensembl database to use.
Default is 103.</p></dd>


<dt id="arg-mirror">mirror<a class="anchor" aria-label="anchor" href="#arg-mirror"></a></dt>
<dd><p>URL of the mirror to use for Ensembl database.
Default is NULL.</p></dd>


<dt id="arg-gtf">gtf<a class="anchor" aria-label="anchor" href="#arg-gtf"></a></dt>
<dd><p>Path to the GTF file to be used for annotation.
Default is NULL.</p></dd>


<dt id="arg-merge-gtf-by">merge_gtf_by<a class="anchor" aria-label="anchor" href="#arg-merge-gtf-by"></a></dt>
<dd><p>Column name to merge the GTF file by.
Default is "gene_name".</p></dd>


<dt id="arg-columns">columns<a class="anchor" aria-label="anchor" href="#arg-columns"></a></dt>
<dd><p>Vector of column names to be used from the GTF file.
Default is "seqname", "feature", "start", "end", "strand", "gene_id", "gene_name", "gene_type".</p></dd>


<dt id="arg-assays">assays<a class="anchor" aria-label="anchor" href="#arg-assays"></a></dt>
<dd><p>Character vector of assay names to be annotated.
Default is "RNA".</p></dd>


<dt id="arg-overwrite">overwrite<a class="anchor" aria-label="anchor" href="#arg-overwrite"></a></dt>
<dd><p>Logical value indicating whether to overwrite existing metadata.
Default is FALSE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="PrepareDB.html">PrepareDB</a></code> <code><a href="ListDB.html">ListDB</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pancreas_sub</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu">AnnotateFeatures</span><span class="op">(</span></span></span>
<span class="r-in"><span>  srt <span class="op">=</span> <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>  species <span class="op">=</span> <span class="st">"Mus_musculus"</span>,</span></span>
<span class="r-in"><span>  db <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"Chromosome"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"GeneType"</span>,</span></span>
<span class="r-in"><span>    <span class="st">"Enzyme"</span>,</span></span>
<span class="r-in"><span>    <span class="co"># "TF",</span></span></span>
<span class="r-in"><span>    <span class="co"># "CSPA",</span></span></span>
<span class="r-in"><span>    <span class="st">"VerSeDa"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:15] Species: Mus_musculus</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:16] Preparing database: Chromosome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:17] Preparing database: GeneType</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:18] Preparing database: Enzyme</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:19] Preparing database: VerSeDa</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:24] Connect to the Ensembl archives...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:25] Using the 103 version of biomart...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:25] Connecting to the biomart...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:27] Searching the dataset mmusculus ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:27] Connecting to the dataset mmusculus_gene_ensembl ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:28] Converting the geneIDs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:31] 11269 genes mapped with ensembl_peptide_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:33] ==============================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:04:33] 11269 genes mapped</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:04:33] 1197 genes unmapped</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:04:33] ==============================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:37] Convert ID types for the database: Chromosome</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:37] Connect to the Ensembl archives...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:38] Using the 103 version of biomart...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:38] Connecting to the biomart...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:38] Searching the dataset mmusculus ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:38] Connecting to the dataset mmusculus_gene_ensembl ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:39] Converting the geneIDs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:53] 27645 genes mapped with entrez_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:04:53] ==============================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:04:53] 27645 genes mapped</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:04:53] 83811 genes unmapped</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:04:53] ==============================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:15] Convert ID types for the database: GeneType</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:15] Connect to the Ensembl archives...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:15] Using the 103 version of biomart...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:15] Connecting to the biomart...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:16] Searching the dataset mmusculus ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:16] Connecting to the dataset mmusculus_gene_ensembl ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:17] Converting the geneIDs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:32] 27645 genes mapped with entrez_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:32] ==============================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:05:32] 27645 genes mapped</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:05:32] 84660 genes unmapped</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:05:32] ==============================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:54] Convert ID types for the database: Enzyme</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:54] Connect to the Ensembl archives...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:54] Using the 103 version of biomart...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:54] Connecting to the biomart...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:56] Searching the dataset mmusculus ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:56] Connecting to the dataset mmusculus_gene_ensembl ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:57] Converting the geneIDs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:57] 2263 genes mapped with entrez_id</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-07-08 15:05:57] ==============================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:05:57] 2263 genes mapped</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:05:57] 20 genes unmapped</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   [2025-07-08 15:05:57] ==============================</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="GetFeaturesData.html">GetFeaturesData</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>    assays <span class="op">=</span> <span class="st">"RNA"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                 TF CSPA Chromosome       GeneType Enzyme          VerSeDa</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Xkr4          &lt;NA&gt; &lt;NA&gt;       chr1 protein-coding   &lt;NA&gt;             &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Mrpl15        &lt;NA&gt; &lt;NA&gt;       chr1 protein-coding   &lt;NA&gt; SecretoryProtein</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4732440D04Rik &lt;NA&gt; &lt;NA&gt;       &lt;NA&gt;           &lt;NA&gt;   &lt;NA&gt;             &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Gm26901       &lt;NA&gt; &lt;NA&gt;       &lt;NA&gt;           &lt;NA&gt;   &lt;NA&gt;             &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Sntg1         &lt;NA&gt; &lt;NA&gt;       chr1 protein-coding   &lt;NA&gt; SecretoryProtein</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Mybl1           TF &lt;NA&gt;       chr1 protein-coding   &lt;NA&gt;             &lt;NA&gt;</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Annotate features using a GTF file</span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu">AnnotateFeatures</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>  gtf <span class="op">=</span> <span class="st">"/refdata-gex-mm10-2020-A/genes/genes.gtf"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="GetFeaturesData.html">GetFeaturesData</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>    assays <span class="op">=</span> <span class="st">"RNA"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hao Zhang, Xu Meng.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

