<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Prefetch cell cycle genes — CycGenePrefetch • scop</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Prefetch cell cycle genes — CycGenePrefetch"><meta name="description" content="Based on the human cell cycle genes,
the cell cycle genes of the corresponding species were captured by homologous gene conversion."><meta property="og:description" content="Based on the human cell cycle genes,
the cell cycle genes of the corresponding species were captured by homologous gene conversion."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scop</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/mengxu98/scop" title="GitHub" aria-label="GitHub repository"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mengxu98" title="MengXu" aria-label="MengXu's GitHub profile"><span class="fa fas fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Prefetch cell cycle genes</h1>
      <small class="dont-index">Source: <a href="https://github.com/mengxu98/scop/blob/dev/R/CycGenePrefetch.R" class="external-link"><code>R/CycGenePrefetch.R</code></a></small>
      <div class="d-none name"><code>CycGenePrefetch.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Based on the human cell cycle genes,
the cell cycle genes of the corresponding species were captured by homologous gene conversion.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CycGenePrefetch</span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="st">"Homo_sapiens"</span>,</span>
<span>  Ensembl_version <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mirror <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  max_tries <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  use_cached_gene <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-species">species<a class="anchor" aria-label="anchor" href="#arg-species"></a></dt>
<dd><p>Latin names for animals, i.e., <code>"Homo_sapiens"</code>, <code>"Mus_musculus"</code></p></dd>


<dt id="arg-ensembl-version">Ensembl_version<a class="anchor" aria-label="anchor" href="#arg-ensembl-version"></a></dt>
<dd><p>Ensembl database version. If NULL, use the current release version.</p></dd>


<dt id="arg-mirror">mirror<a class="anchor" aria-label="anchor" href="#arg-mirror"></a></dt>
<dd><p>Specify an Ensembl mirror to connect to. The valid options here are <code>"www"</code>, <code>"uswest"</code>, <code>"useast"</code>, <code>"asia"</code>.</p></dd>


<dt id="arg-max-tries">max_tries<a class="anchor" aria-label="anchor" href="#arg-max-tries"></a></dt>
<dd><p>The maximum number of attempts to connect with the BioMart service.</p></dd>


<dt id="arg-use-cached-gene">use_cached_gene<a class="anchor" aria-label="anchor" href="#arg-use-cached-gene"></a></dt>
<dd><p>Whether to use previously cached cell cycle gene conversion results for the species.
Default is <code>TRUE</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Whether to print the message.
Default is <code>TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list of S-phase and G2M-phase genes.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="GeneConvert.html">GeneConvert</a></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">ccgenes</span> <span class="op">&lt;-</span> <span class="fu">CycGenePrefetch</span><span class="op">(</span><span class="st">"Homo_sapiens"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:25] Prefetching cell cycle genes for <span style="color: #0000BB;">"Homo_sapiens"</span> ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> [2025-09-20 13:08:25] Cell cycle gene prefetching completed <span style="color: #0000BB;">"Homo_sapiens"</span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">ccgenes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ res: NULL</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ S  : chr [1:43] "MCM5" "PCNA" "TYMS" "FEN1" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ G2M: chr [1:54] "HMGB2" "CDK1" "NUSAP1" "UBE2C" ...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">ccgenes</span> <span class="op">&lt;-</span> <span class="fu">CycGenePrefetch</span><span class="op">(</span><span class="st">"Mus_musculus"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:25] Prefetching cell cycle genes for <span style="color: #0000BB;">"Mus_musculus"</span> ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:25] Installing: <span style="color: #0000BB;">biomaRt</span>...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> → Will <span style="font-style: italic;">install</span> 10 packages.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> → All 10 packages (0 B) are cached.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">+ </span><span style="color: #0000BB;">BiocFileCache</span>   2.16.2 [bld]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">+ </span><span style="color: #0000BB;">biomaRt</span>         2.64.0 [bld]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">+ </span><span style="color: #0000BB;">dbplyr</span>          2.5.1  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">+ </span><span style="color: #0000BB;">filelock</span>        1.0.3  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">+ </span><span style="color: #0000BB;">hms</span>             1.1.3  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">+ </span><span style="color: #0000BB;">httr2</span>           1.2.1  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">+ </span><span style="color: #0000BB;">prettyunits</span>     1.2.0  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">+ </span><span style="color: #0000BB;">progress</span>        1.2.3  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">+ </span><span style="color: #0000BB;">tidyr</span>           1.3.1  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">+ </span><span style="color: #0000BB;">xml2</span>            1.4.0  <span style="color: #555555;"> + </span><span style="color: #00BB00;">✔</span><span style="color: #00BBBB;"> libxml2-dev</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> All system requirements are already installed.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>   </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> No downloads are needed, 10 pkgs are cached</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Got <span style="color: #0000BB;">filelock</span> 1.0.3 (x86_64-pc-linux-gnu-ubuntu-24.04) (24.70 kB)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Got <span style="color: #0000BB;">hms</span> 1.1.3 (x86_64-pc-linux-gnu-ubuntu-24.04) (99.70 kB)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Got <span style="color: #0000BB;">progress</span> 1.2.3 (x86_64-pc-linux-gnu-ubuntu-24.04) (87.00 kB)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Got <span style="color: #0000BB;">prettyunits</span> 1.2.0 (x86_64-pc-linux-gnu-ubuntu-24.04) (152.62 kB)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Got <span style="color: #0000BB;">xml2</span> 1.4.0 (x86_64-pc-linux-gnu-ubuntu-24.04) (280.17 kB)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Got <span style="color: #0000BB;">BiocFileCache</span> 2.16.2 (source) (740.08 kB)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Got <span style="color: #0000BB;">biomaRt</span> 2.64.0 (source) (862.68 kB)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Got <span style="color: #0000BB;">httr2</span> 1.2.1 (x86_64-pc-linux-gnu-ubuntu-24.04) (782.87 kB)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Got <span style="color: #0000BB;">tidyr</span> 1.3.1 (x86_64-pc-linux-gnu-ubuntu-24.04) (1.18 MB)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Got <span style="color: #0000BB;">dbplyr</span> 2.5.1 (x86_64-pc-linux-gnu-ubuntu-24.04) (1.24 MB)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Installing system requirements</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Executing `sudo sh -c apt-get -y update`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Get:1 file:/etc/apt/apt-mirrors.txt Mirrorlist [144 B]</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Hit:2 http://azure.archive.ubuntu.com/ubuntu noble InRelease</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Hit:3 http://azure.archive.ubuntu.com/ubuntu noble-updates InRelease</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Hit:4 http://azure.archive.ubuntu.com/ubuntu noble-backports InRelease</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Hit:5 http://azure.archive.ubuntu.com/ubuntu noble-security InRelease</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Hit:6 https://packages.microsoft.com/repos/azure-cli noble InRelease</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Hit:7 https://packages.microsoft.com/ubuntu/24.04/prod noble InRelease</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading package lists...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Executing `sudo sh -c apt-get -y install libxml2-dev libcurl4-openssl-dev libssl-dev libpng-dev libicu-dev`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading package lists...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Building dependency tree...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reading state information...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> libxml2-dev is already the newest version (2.9.14+dfsg-1.3ubuntu3.5).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> libxml2-dev set to manually installed.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> libcurl4-openssl-dev is already the newest version (8.5.0-2ubuntu10.6).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> libssl-dev is already the newest version (3.0.13-0ubuntu3.5).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> libpng-dev is already the newest version (1.6.43-5build1).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> libicu-dev is already the newest version (74.2-1ubuntu3.1).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 0 upgraded, 0 newly installed, 0 to remove and 41 not upgraded.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Installed <span style="color: #0000BB;">dbplyr</span> 2.5.1  <span style="color: #9E9E9E;">(63ms)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Installed <span style="color: #0000BB;">filelock</span> 1.0.3  <span style="color: #9E9E9E;">(78ms)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Installed <span style="color: #0000BB;">hms</span> 1.1.3  <span style="color: #9E9E9E;">(99ms)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Installed <span style="color: #0000BB;">httr2</span> 1.2.1  <span style="color: #9E9E9E;">(130ms)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Installed <span style="color: #0000BB;">prettyunits</span> 1.2.0  <span style="color: #9E9E9E;">(60ms)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Installed <span style="color: #0000BB;">progress</span> 1.2.3  <span style="color: #9E9E9E;">(103ms)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Installed <span style="color: #0000BB;">tidyr</span> 1.3.1  <span style="color: #9E9E9E;">(68ms)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Building <span style="color: #0000BB;">BiocFileCache</span> 2.16.2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Installed <span style="color: #0000BB;">xml2</span> 1.4.0  <span style="color: #9E9E9E;">(77ms)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Built <span style="color: #0000BB;">BiocFileCache</span> 2.16.2 <span style="color: #9E9E9E;">(3.8s)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Installed <span style="color: #0000BB;">BiocFileCache</span> 2.16.2  <span style="color: #9E9E9E;">(32ms)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Building <span style="color: #0000BB;">biomaRt</span> 2.64.0</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Built <span style="color: #0000BB;">biomaRt</span> 2.64.0 <span style="color: #9E9E9E;">(7.8s)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Installed <span style="color: #0000BB;">biomaRt</span> 2.64.0  <span style="color: #9E9E9E;">(1s)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> 1 pkg + 57 deps: kept 48, added 10, dld 10 (5.45 MB) <span style="color: #B2B2B2;">[17.4s]</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:43] <span style="color: #0000BB;">biomaRt</span> installed successfully</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:43] Connect to the Ensembl archives...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:44] Using the <span style="color: #0000BB;">"115"</span> version of biomart...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:44] Downloading the biomart from <span style="color: #0000BB;">"https://sep2025.archive.ensembl.org"</span>...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:46] Searching the dataset <span style="color: #0000BB;">"hsapiens"</span> ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:47] Connecting to the dataset <span style="color: #0000BB;">hsapiens_gene_ensembl</span> ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:48] Converting the geneIDs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:49] <span style="color: #0000BB;">97</span> genes mapped with <span style="color: #0000BB;">"ensembl_symbol"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:49] ==============================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span>                       <span style="color: #00BB00;">97 genes mapped</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span>                       <span style="color: #BB0000;">0 genes unmapped</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span>                       ==============================</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2025-09-20 13:08:49] Cached conversion results for <span style="color: #0000BB;">"Mus_musculus"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> [2025-09-20 13:08:49] Cell cycle gene prefetching completed <span style="color: #0000BB;">"Mus_musculus"</span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">ccgenes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ res:List of 7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ geneID_res     :'data.frame':	100 obs. of  4 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ from_IDtype: chr [1:100] "ensembl_symbol" "ensembl_symbol" "ensembl_symbol" "ensembl_symbol" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ from_geneID: chr [1:100] "NCAPD2" "ANLN" "UBR7" "TACC3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ to_IDtype  : chr [1:100] "symbol" "symbol" "symbol" "symbol" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ to_geneID  : chr [1:100] "Ncapd2" "Anln" "Ubr7" "Tacc3" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ geneID_collapse:'data.frame':	96 obs. of  2 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ from_geneID: chr [1:96] "ANLN" "ANP32E" "ATAD2" "AURKA" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ symbol     :List of 96</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Anln"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Anp32e"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Atad2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Aurka"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Aurkb"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Birc5"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Blm"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Bub1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Casp8ap2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cbx5"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ccnb2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ccne2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cdc20"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cdc25c"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cdc45"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cdc6"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cdca2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cdca3"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cdca7"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cdca8"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cdk1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cenpa"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cenpe"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cenpf"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cenpu"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Chaf1b"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ckap2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ckap2l"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ckap5"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr [1:2] "Cks1brt" "Cks1b"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Cks2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Clspn"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ctcf"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Dlgap5"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Dscc1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Dtl"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "E2f8"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ect2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Exo1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Fen1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "G2e3"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Gas2l3"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Gins2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Gmnn"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Gtse1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Hells"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Hjurp"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Hmgb2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Hmmr"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Jpt1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Kif11"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Kif20b"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Kif23"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Kif2c"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Lbr"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Mcm4"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Mcm5"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Mcm6"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Mcm7"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Mki67"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Mrpl36"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Msh2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Nasp"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ncapd2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ndc80"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Nek2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Nuf2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Nusap1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Pcna"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Pimreg"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Pola1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Polr1b"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Prim1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Psrc1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Rad51"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Rad51ap1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Rangap1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Rfc2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Rrm1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Rrm2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Slbp"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Smc4"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Tacc3"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Tipin"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Tmpo"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Top2a"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Tpx2"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ttk"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Tubb4b"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Tyms"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ube2c"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ubr7"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Uhrf1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Ung"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Usp1"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..$ : chr "Wdr76"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. .. ..- attr(*, "class")= chr "AsIs"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ geneID_expand  :'data.frame':	97 obs. of  2 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ from_geneID: chr [1:97] "ANLN" "ANP32E" "ATAD2" "AURKA" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ symbol     : chr [1:97] "Anln" "Anp32e" "Atad2" "Aurka" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ Ensembl_version: chr "115"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ Datasets       :'data.frame':	213 obs. of  3 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ dataset    : 'AsIs' chr [1:213] "abrachyrhynchus_gene_ensembl" "acalliptera_gene_ensembl" "acarolinensis_gene_ensembl" "acchrysaetos_gene_ensembl" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ description: 'AsIs' chr [1:213] "Pink-footed goose genes (ASM259213v1)" "Eastern happy genes (fAstCal1.3)" "Green anole genes (AnoCar2.0v2)" "Golden eagle genes (bAquChr1.2)" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ version    : 'AsIs' chr [1:213] "ASM259213v1" "fAstCal1.3" "AnoCar2.0v2" "bAquChr1.2" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ Attributes     :'data.frame':	3170 obs. of  3 variables:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ name       : chr [1:3170] "ensembl_gene_id" "ensembl_gene_id_version" "ensembl_transcript_id" "ensembl_transcript_id_version" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ description: chr [1:3170] "Gene stable ID" "Gene stable ID version" "Transcript stable ID" "Transcript stable ID version" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   .. ..$ page       : chr [1:3170] "feature_page" "feature_page" "feature_page" "feature_page" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ..$ geneID_unmapped: chr(0) </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ S  : chr [1:42] "Mcm5" "Pcna" "Tyms" "Fen1" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ G2M: chr [1:55] "Hmgb2" "Cdk1" "Nusap1" "Ube2c" ...</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xu Meng, Hao Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

