<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Find Expressed Markers — FindExpressedMarkers • scop</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Find Expressed Markers — FindExpressedMarkers"><meta name="description" content="Find Expressed Markers"><meta property="og:description" content="Find Expressed Markers"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scop</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/mengxu98/scop" title="GitHub" aria-label="GitHub repository"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mengxu98" title="MengXu" aria-label="MengXu's GitHub profile"><span class="fa fas fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Find Expressed Markers</h1>
      <small class="dont-index">Source: <a href="https://github.com/mengxu98/scop/blob/dev/R/FindExpressedMarkers.R" class="external-link"><code>R/FindExpressedMarkers.R</code></a></small>
      <div class="d-none name"><code>FindExpressedMarkers.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Find Expressed Markers</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">FindExpressedMarkers</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  ident.1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ident.2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cells.1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cells.2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  features <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  layer <span class="op">=</span> <span class="st">"data"</span>,</span>
<span>  min.expression <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  test.use <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>  logfc.threshold <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  base <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  pseudocount.use <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  mean.fxn <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fc.name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min.pct <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  min.diff.pct <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>,</span>
<span>  max.cells.per.ident <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  latent.vars <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  only.pos <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  min.cells.group <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  min.cells.feature <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  norm.method <span class="op">=</span> <span class="st">"LogNormalize"</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">11</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>An object</p></dd>


<dt id="arg-ident-">ident.1<a class="anchor" aria-label="anchor" href="#arg-ident-"></a></dt>
<dd><p>Identity class to define markers for; pass an object of class
<code>phylo</code> or 'clustertree' to find markers for a node in a cluster tree;
passing 'clustertree' requires <code><a href="https://satijalab.org/seurat/reference/BuildClusterTree.html" class="external-link">BuildClusterTree</a></code> to have been run</p></dd>


<dt id="arg-ident-">ident.2<a class="anchor" aria-label="anchor" href="#arg-ident-"></a></dt>
<dd><p>A second identity class for comparison; if <code>NULL</code>,
use all other cells for comparison; if an object of class <code>phylo</code> or
'clustertree' is passed to <code>ident.1</code>, must pass a node to find markers for</p></dd>


<dt id="arg-cells-">cells.1<a class="anchor" aria-label="anchor" href="#arg-cells-"></a></dt>
<dd><p>Vector of cell names belonging to group 1</p></dd>


<dt id="arg-cells-">cells.2<a class="anchor" aria-label="anchor" href="#arg-cells-"></a></dt>
<dd><p>Vector of cell names belonging to group 2</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>A character vector or a named list of features to plot.
Features can be gene names in Assay or names of numeric columns in meta.data.</p></dd>


<dt id="arg-assay">assay<a class="anchor" aria-label="anchor" href="#arg-assay"></a></dt>
<dd><p>Which assay to use.
If <code>NULL</code>, the default assay of the Seurat object will be used.</p></dd>


<dt id="arg-layer">layer<a class="anchor" aria-label="anchor" href="#arg-layer"></a></dt>
<dd><p>Which layer to use.
Default is <code>data</code>.</p></dd>


<dt id="arg-min-expression">min.expression<a class="anchor" aria-label="anchor" href="#arg-min-expression"></a></dt>
<dd><p>The min.expression used.</p></dd>


<dt id="arg-test-use">test.use<a class="anchor" aria-label="anchor" href="#arg-test-use"></a></dt>
<dd><p>Denotes which test to use. Available options are:</p><ul><li><p>"wilcox" : Identifies differentially expressed genes between two
 groups of cells using a Wilcoxon Rank Sum test (default); will use a fast
 implementation by Presto if installed</p></li>
<li><p>"wilcox_limma" : Identifies differentially expressed genes between two
 groups of cells using the limma implementation of the Wilcoxon Rank Sum test;
 set this option to reproduce results from Seurat v4</p></li>
<li><p>"bimod" : Likelihood-ratio test for single cell gene expression,
 (McDavid et al., Bioinformatics, 2013)</p></li>
<li><p>"roc" : Identifies 'markers' of gene expression using ROC analysis.
 For each gene, evaluates (using AUC) a classifier built on that gene alone,
 to classify between two groups of cells. An AUC value of 1 means that
 expression values for this gene alone can perfectly classify the two
 groupings (i.e. Each of the cells in cells.1 exhibit a higher level than
 each of the cells in cells.2). An AUC value of 0 also means there is perfect
 classification, but in the other direction. A value of 0.5 implies that
 the gene has no predictive power to classify the two groups. Returns a
 'predictive power' (abs(AUC-0.5) * 2) ranked matrix of putative differentially
 expressed genes.</p></li>
<li><p>"t" : Identify differentially expressed genes between two groups of
 cells using the Student's t-test.</p></li>
<li><p>"negbinom" : Identifies differentially expressed genes between two
  groups of cells using a negative binomial generalized linear model.
  Use only for UMI-based datasets</p></li>
<li><p>"poisson" : Identifies differentially expressed genes between two
  groups of cells using a poisson generalized linear model.
  Use only for UMI-based datasets</p></li>
<li><p>"LR" : Uses a logistic regression framework to determine differentially
 expressed genes. Constructs a logistic regression model predicting group
 membership based on each feature individually and compares this to a null
 model with a likelihood ratio test.</p></li>
<li><p>"MAST" : Identifies differentially expressed genes between two groups
 of cells using a hurdle model tailored to scRNA-seq data. Utilizes the MAST
 package to run the DE testing.</p></li>
<li><p>"DESeq2" : Identifies differentially expressed genes between two groups
 of cells based on a model using DESeq2 which uses a negative binomial
 distribution (Love et al, Genome Biology, 2014).This test does not support
 pre-filtering of genes based on average difference (or percent detection rate)
 between cell groups. However, genes may be pre-filtered based on their
 minimum detection rate (min.pct) across both cell groups. To use this method,
 please install DESeq2, using the instructions at
 https://bioconductor.org/packages/release/bioc/html/DESeq2.html</p></li>
</ul></dd>


<dt id="arg-logfc-threshold">logfc.threshold<a class="anchor" aria-label="anchor" href="#arg-logfc-threshold"></a></dt>
<dd><p>Limit testing to genes which show, on average, at least
X-fold difference (log-scale) between the two groups of cells. Default is 0.1
Increasing logfc.threshold speeds up the function, but can miss weaker signals.
If the <code>slot</code> parameter is "scale.data" no filtering is performed.</p></dd>


<dt id="arg-base">base<a class="anchor" aria-label="anchor" href="#arg-base"></a></dt>
<dd><p>The base with respect to which logarithms are computed.</p></dd>


<dt id="arg-pseudocount-use">pseudocount.use<a class="anchor" aria-label="anchor" href="#arg-pseudocount-use"></a></dt>
<dd><p>Pseudocount to add to averaged expression values when
calculating logFC. 1 by default.</p></dd>


<dt id="arg-mean-fxn">mean.fxn<a class="anchor" aria-label="anchor" href="#arg-mean-fxn"></a></dt>
<dd><p>Function to use for fold change or average difference calculation.
The default depends on the the value of <code>fc.slot</code>:</p><ul><li><p>"counts" : difference in the log of the mean counts, with pseudocount.</p></li>
<li><p>"data" : difference in the log of the average exponentiated data, with pseudocount.
 This adjusts for differences in sequencing depth between cells, and assumes that "data"
 has been log-normalized.</p></li>
<li><p>"scale.data" : difference in the means of scale.data.</p></li>
</ul></dd>


<dt id="arg-fc-name">fc.name<a class="anchor" aria-label="anchor" href="#arg-fc-name"></a></dt>
<dd><p>Name of the fold change, average difference, or custom function column
in the output data.frame. If NULL, the fold change column will be named
according to the logarithm base (eg, "avg_log2FC"), or if using the scale.data
slot "avg_diff".</p></dd>


<dt id="arg-min-pct">min.pct<a class="anchor" aria-label="anchor" href="#arg-min-pct"></a></dt>
<dd><p>only test genes that are detected in a minimum fraction of
min.pct cells in either of the two populations. Meant to speed up the function
by not testing genes that are very infrequently expressed. Default is 0.01</p></dd>


<dt id="arg-min-diff-pct">min.diff.pct<a class="anchor" aria-label="anchor" href="#arg-min-diff-pct"></a></dt>
<dd><p>only test genes that show a minimum difference in the
fraction of detection between the two groups. Set to -Inf by default</p></dd>


<dt id="arg-max-cells-per-ident">max.cells.per.ident<a class="anchor" aria-label="anchor" href="#arg-max-cells-per-ident"></a></dt>
<dd><p>Down sample each identity class to a max number.
Default is no downsampling. Not activated by default (set to Inf)</p></dd>


<dt id="arg-latent-vars">latent.vars<a class="anchor" aria-label="anchor" href="#arg-latent-vars"></a></dt>
<dd><p>Variables to test, used only when <code>test.use</code> is one of
'LR', 'negbinom', 'poisson', or 'MAST'</p></dd>


<dt id="arg-only-pos">only.pos<a class="anchor" aria-label="anchor" href="#arg-only-pos"></a></dt>
<dd><p>Only return positive markers (FALSE by default)</p></dd>


<dt id="arg-min-cells-group">min.cells.group<a class="anchor" aria-label="anchor" href="#arg-min-cells-group"></a></dt>
<dd><p>Minimum number of cells in one of the groups</p></dd>


<dt id="arg-min-cells-feature">min.cells.feature<a class="anchor" aria-label="anchor" href="#arg-min-cells-feature"></a></dt>
<dd><p>Minimum number of cells expressing the feature in at least one
of the two groups, currently only used for poisson and negative binomial tests</p></dd>


<dt id="arg-norm-method">norm.method<a class="anchor" aria-label="anchor" href="#arg-norm-method"></a></dt>
<dd><p>Normalization method for fold change calculation when
<code>slot</code> is “<code>data</code>”</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Random seed for reproducibility.
Default is <code>11</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Whether to print the message.
Default is <code>TRUE</code>.</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Arguments passed to other methods and to specific DE methods</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>[RunDEtest]</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pancreas_sub</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="standard_scop.html">standard_scop</a></span><span class="op">(</span><span class="va">pancreas_sub</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:04] Start standard scop workflow...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:04] Checking a list of <span style="color: #0000BB;">&lt;Seurat&gt;</span>...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> [2026-01-27 07:51:04] Data <span style="color: #0000BB;">1</span>/<span style="color: #0000BB;">1</span> of the `srt_list` is <span style="color: #0000BB;">"unknown"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:04] Perform `NormalizeData()` with `normalization.method = 'LogNormalize'` on the data <span style="color: #0000BB;">1</span>/<span style="color: #0000BB;">1</span> of the `srt_list`...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:07] Perform `Seurat::FindVariableFeatures()` on the data <span style="color: #0000BB;">1</span>/<span style="color: #0000BB;">1</span> of the `srt_list`...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:07] Use the separate HVF from srt_list</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:07] Number of available HVF: <span style="color: #0000BB;">2000</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:07] Finished check</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:08] Perform `Seurat::ScaleData()`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:08] Perform <span style="color: #0000BB;">pca</span> linear dimension reduction</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:09] Perform `Seurat::FindClusters()` with `cluster_algorithm = 'louvain'` and `cluster_resolution = 0.6`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:09] Reorder clusters...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:09] Perform <span style="color: #0000BB;">umap</span> nonlinear dimension reduction</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:09] Non-linear dimensionality reduction (<span style="color: #0000BB;">umap</span>) using (<span style="color: #0000BB;">Standardpca</span>) dims (<span style="color: #0000BB;">1</span>-<span style="color: #0000BB;">50</span>) as input</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:13] Non-linear dimensionality reduction (<span style="color: #0000BB;">umap</span>) using (<span style="color: #0000BB;">Standardpca</span>) dims (<span style="color: #0000BB;">1</span>-<span style="color: #0000BB;">50</span>) as input</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> [2026-01-27 07:51:16] Run scop standard workflow completed</span>
<span class="r-in"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">FindExpressedMarkers</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>  cells.1 <span class="op">=</span> <span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/WhichCells.html" class="external-link">WhichCells</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>    expression <span class="op">=</span> <span class="va">Phase</span> <span class="op">==</span> <span class="st">"G2M"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:17] Using <span style="color: #0000BB;">1</span> core</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #FFAF00;">⠙</span> [2026-01-27 07:51:17] Running for <span style="color: #0000BB;">1 </span>[<span style="color: #0000BB;">1</span>/<span style="color: #0000BB;">6145</span>] <span style="color: #00BB00;">■                               </span>…</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #FFAF00;">⠹</span> [2026-01-27 07:51:17] Running for <span style="color: #0000BB;">3591 </span>[<span style="color: #0000BB;">3591</span>/<span style="color: #0000BB;">6145</span>] <span style="color: #00BB00;">■■■■■■■■■■■■■■■■■■■       </span>…</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> [2026-01-27 07:51:17] Completed <span style="color: #0000BB;">6145</span> tasks in <span style="color: #0000BB;">3.4s</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> [2026-01-27 07:51:17] Building results</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               p_val avg_log2FC pct.1 pct.2    p_val_adj</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Hmgb2  1.964385e-37  2.1638592 1.000 0.526 3.142624e-33</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Tuba1b 3.652521e-33  1.8756996 0.986 0.801 5.843303e-29</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Ran    2.573020e-31  1.2068807 1.000 0.946 4.116317e-27</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> H2afx  9.845758e-31  1.4231361 0.971 0.519 1.575124e-26</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Ptma   3.541885e-30  0.9021115 1.000 0.999 5.666307e-26</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Tubb5  1.550857e-29  1.4861586 1.000 0.940 2.481062e-25</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="FeatureStatPlot.html">FeatureStatPlot</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>  group.by <span class="op">=</span> <span class="st">"Phase"</span>,</span></span>
<span class="r-in"><span>  add_point <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>No shared levels found between `names(values)` of the manual scale and the</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> data's <span style="color: #00BB00;">colour</span> values.</span>
<span class="r-plt img"><img src="FindExpressedMarkers-1.png" alt="" width="700" height="433"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xu Meng, Hao Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

