<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ scop</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scop</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/mengxu98/scop" title="GitHub" aria-label="GitHub repository"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mengxu98" title="MengXu" aria-label="MengXu's GitHub profile"><span class="fa fas fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">

      <h3 id="single-cell-omics-analysis-pipeline">Single-Cell Omics analysis Pipeline<a class="anchor" aria-label="anchor" href="#single-cell-omics-analysis-pipeline"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="scop-package.html">scop</a></code> <code><a href="scop-package.html">scop-package</a></code>

        </dt>
        <dd>Single-Cell Omics analysis Pipeline</dd>

        <dt>

          <code><a href="scop_logo.html">scop_logo()</a></code>

        </dt>
        <dd>scop logo</dd>
      </dl></div><div class="section level2">

      <h3 id="package-management">Package Management<a class="anchor" aria-label="anchor" href="#package-management"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="check_python.html">check_python()</a></code>

        </dt>
        <dd>Check and install python packages</dd>

        <dt>

          <code><a href="check_r.html">check_r()</a></code>

        </dt>
        <dd>Check and install R packages</dd>

        <dt>

          <code><a href="conda_install.html">conda_install()</a></code>

        </dt>
        <dd>Enhanced conda installation</dd>

        <dt>

          <code><a href="conda_python.html">conda_python()</a></code>

        </dt>
        <dd>Find the path to Python associated with a conda environment</dd>

        <dt>

          <code><a href="env_exist.html">env_exist()</a></code>

        </dt>
        <dd>Check if a conda environment exists</dd>

        <dt>

          <code><a href="env_info.html">env_info()</a></code>

        </dt>
        <dd>Print environment information</dd>

        <dt>

          <code><a href="env_requirements.html">env_requirements()</a></code>

        </dt>
        <dd>Python environment requirements</dd>

        <dt>

          <code><a href="exist_python_pkgs.html">exist_python_pkgs()</a></code>

        </dt>
        <dd>Check if the python package exists in the environment</dd>

        <dt>

          <code><a href="find_conda.html">find_conda()</a></code>

        </dt>
        <dd>Find an appropriate conda binary</dd>

        <dt>

          <code><a href="install_miniconda2.html">install_miniconda2()</a></code>

        </dt>
        <dd>Enhanced miniconda installation</dd>

        <dt>

          <code><a href="installed_python_pkgs.html">installed_python_pkgs()</a></code>

        </dt>
        <dd>Show all the python packages in the environment</dd>

        <dt>

          <code><a href="ListEnv.html">ListEnv()</a></code>

        </dt>
        <dd>List conda environments</dd>

        <dt>

          <code><a href="PrepareEnv.html">PrepareEnv()</a></code>

        </dt>
        <dd>Prepare the virtual environment</dd>

        <dt>

          <code><a href="remove_python.html">remove_python()</a></code>

        </dt>
        <dd>Remove Python packages from conda environment</dd>

        <dt>

          <code><a href="remove_r.html">remove_r()</a></code>

        </dt>
        <dd>Check and remove R packages</dd>

        <dt>

          <code><a href="RemoveEnv.html">RemoveEnv()</a></code>

        </dt>
        <dd>Remove a conda environment</dd>
      </dl></div><div class="section level2">

      <h3 id="scop-pipeline">scop pipeline<a class="anchor" aria-label="anchor" href="#scop-pipeline"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="integration_scop.html">integration_scop()</a></code>

        </dt>
        <dd>The integration_scop function</dd>

        <dt>

          <code><a href="standard_scop.html">standard_scop()</a></code>

        </dt>
        <dd>Standard workflow for scop</dd>
      </dl></div><div class="section level2">

      <h3 id="data-processing">Data Processing<a class="anchor" aria-label="anchor" href="#data-processing"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="adata_to_srt.html">adata_to_srt()</a></code>

        </dt>
        <dd>Convert an anndata object to a seurat object using reticulate</dd>

        <dt>

          <code><a href="CellScoring.html">CellScoring()</a></code>

        </dt>
        <dd>Cell scoring</dd>

        <dt>

          <code><a href="CheckDataType.html">CheckDataType()</a></code>

        </dt>
        <dd>Check and report the type of data in <code>Seurat</code> object</dd>

        <dt>

          <code><a href="CheckDataList.html">CheckDataList()</a></code>

        </dt>
        <dd>Check and preprocess a list of <code>Seurat</code> objects</dd>

        <dt>

          <code><a href="CheckDataMerge.html">CheckDataMerge()</a></code>

        </dt>
        <dd>Check and preprocess a merged seurat object</dd>

        <dt>

          <code><a href="CycGenePrefetch.html">CycGenePrefetch()</a></code>

        </dt>
        <dd>Prefetch cell cycle genes</dd>

        <dt>

          <code><a href="DefaultReduction.html">DefaultReduction()</a></code>

        </dt>
        <dd>Find the default reduction name in a Seurat object</dd>

        <dt>

          <code><a href="FetchDataZero.html">FetchDataZero()</a></code>

        </dt>
        <dd>FetchData but with zeroes for unavailable genes</dd>

        <dt>

          <code><a href="GeneConvert.html">GeneConvert()</a></code>

        </dt>
        <dd>Gene ID conversion function using biomart</dd>

        <dt>

          <code><a href="GetAssayData5.html">GetAssayData5()</a></code>

        </dt>
        <dd>Get expression data from <code>Assay5</code> or Seurat object</dd>

        <dt>

          <code><a href="is_outlier.html">is_outlier()</a></code>

        </dt>
        <dd>Detect outliers using MAD (Median Absolute Deviation)</dd>

        <dt>

          <code><a href="RecoverCounts.html">RecoverCounts()</a></code>

        </dt>
        <dd>Attempt to recover raw counts from the normalized matrix</dd>

        <dt>

          <code><a href="RenameClusters.html">RenameClusters()</a></code>

        </dt>
        <dd>Rename clusters for the Seurat object</dd>

        <dt>

          <code><a href="srt_append.html">srt_append()</a></code>

        </dt>
        <dd>Append a Seurat object to another</dd>

        <dt>

          <code><a href="srt_to_adata.html">srt_to_adata()</a></code>

        </dt>
        <dd>Convert a Seurat object to an AnnData object</dd>

        <dt>

          <code><a href="srt_reorder.html">srt_reorder()</a></code>

        </dt>
        <dd>Reorder idents by the gene expression</dd>
      </dl></div><div class="section level2">

      <h3 id="feature-processing">Feature Processing<a class="anchor" aria-label="anchor" href="#feature-processing"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="AddFeaturesData.html">AddFeaturesData()</a></code>

        </dt>
        <dd>Add features data</dd>

        <dt>

          <code><a href="AnnotateFeatures.html">AnnotateFeatures()</a></code>

        </dt>
        <dd>Annotate Features</dd>

        <dt>

          <code><a href="GetFeaturesData.html">GetFeaturesData()</a></code>

        </dt>
        <dd>Get features data</dd>

        <dt>

          <code><a href="GetSimilarFeatures.html">GetSimilarFeatures()</a></code>

        </dt>
        <dd>Find genes with expression patterns similar to the genes you've specified.</dd>

        <dt>

          <code><a href="RenameFeatures.html">RenameFeatures()</a></code>

        </dt>
        <dd>Rename features for the Seurat object</dd>
      </dl></div><div class="section level2">

      <h3 id="quality-control">Quality Control<a class="anchor" aria-label="anchor" href="#quality-control"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="db_DoubletDetection.html">db_DoubletDetection()</a></code>

        </dt>
        <dd>Run doublet-calling with DoubletDetection</dd>

        <dt>

          <code><a href="db_scDblFinder.html">db_scDblFinder()</a></code>

        </dt>
        <dd>Run doublet-calling with scDblFinder</dd>

        <dt>

          <code><a href="db_scds.html">db_scds()</a></code>

        </dt>
        <dd>Run doublet-calling with scds</dd>

        <dt>

          <code><a href="db_Scrublet.html">db_Scrublet()</a></code>

        </dt>
        <dd>Run doublet-calling with Scrublet</dd>

        <dt>

          <code><a href="RunCellQC.html">RunCellQC()</a></code>

        </dt>
        <dd>Run cell-level quality control for single cell RNA-seq data.</dd>

        <dt>

          <code><a href="RunDoubletCalling.html">RunDoubletCalling()</a></code>

        </dt>
        <dd>Run doublet-calling for single cell RNA-seq data.</dd>
      </dl></div><div class="section level2">

      <h3 id="dimensionality-reduction-and-clustering">Dimensionality Reduction and Clustering<a class="anchor" aria-label="anchor" href="#dimensionality-reduction-and-clustering"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="RunDimReduction.html">RunDimReduction()</a></code>

        </dt>
        <dd>Run dimensionality reduction</dd>

        <dt>

          <code><a href="RunDM.html">RunDM()</a></code>

        </dt>
        <dd>Run DM (diffusion map)</dd>

        <dt>

          <code><a href="RunFR.html">RunFR()</a></code>

        </dt>
        <dd>Run Force-Directed Layout (Fruchterman-Reingold algorithm)</dd>

        <dt>

          <code><a href="RunGLMPCA.html">RunGLMPCA()</a></code>

        </dt>
        <dd>Run GLMPCA (generalized version of principal components analysis)</dd>

        <dt>

          <code><a href="RunMDS.html">RunMDS()</a></code>

        </dt>
        <dd>Run MDS (multi-dimensional scaling)</dd>

        <dt>

          <code><a href="RunNMF.html">RunNMF()</a></code>

        </dt>
        <dd>Run NMF (non-negative matrix factorization)</dd>

        <dt>

          <code><a href="RunUMAP2.html">RunUMAP2()</a></code>

        </dt>
        <dd>Run UMAP (Uniform Manifold Approximation and Projection)</dd>

        <dt>

          <code><a href="RunPaCMAP.html">RunPaCMAP()</a></code>

        </dt>
        <dd>Run PaCMAP (Pairwise Controlled Manifold Approximation)</dd>

        <dt>

          <code><a href="RunTriMap.html">RunTriMap()</a></code>

        </dt>
        <dd>Run TriMap (Large-scale Dimensionality Reduction Using Triplets)</dd>

        <dt>

          <code><a href="RunPHATE.html">RunPHATE()</a></code>

        </dt>
        <dd>Run PHATE (Potential of Heat-diffusion for Affinity-based Trajectory Embedding)</dd>

        <dt>

          <code><a href="RunLargeVis.html">RunLargeVis()</a></code>

        </dt>
        <dd>Run LargeVis (Dimensionality Reduction with a LargeVis-like method)</dd>

        <dt>

          <code><a href="RunKNNPredict.html">RunKNNPredict()</a></code>

        </dt>
        <dd>Run KNN prediction</dd>
      </dl></div><div class="section level2">

      <h3 id="data-integration-and-batch-effect-removal">Data Integration and Batch Effect Removal<a class="anchor" aria-label="anchor" href="#data-integration-and-batch-effect-removal"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="RunCSSMap.html">RunCSSMap()</a></code>

        </dt>
        <dd>Single-cell reference mapping with CSS method</dd>

        <dt>

          <code><a href="RunHarmony2.html">RunHarmony2()</a></code>

        </dt>
        <dd>Run Harmony algorithm</dd>

        <dt>

          <code><a href="RunKNNMap.html">RunKNNMap()</a></code>

        </dt>
        <dd>Single-cell reference mapping with KNN method</dd>

        <dt>

          <code><a href="RunPCAMap.html">RunPCAMap()</a></code>

        </dt>
        <dd>Single-cell reference mapping with PCA method</dd>

        <dt>

          <code><a href="RunScmap.html">RunScmap()</a></code>

        </dt>
        <dd>Annotate single cells using scmap.</dd>

        <dt>

          <code><a href="RunSeuratMap.html">RunSeuratMap()</a></code>

        </dt>
        <dd>Single-cell reference mapping with Seurat method</dd>

        <dt>

          <code><a href="RunSingleR.html">RunSingleR()</a></code>

        </dt>
        <dd>Annotate single cells using SingleR</dd>

        <dt>

          <code><a href="RunSymphonyMap.html">RunSymphonyMap()</a></code>

        </dt>
        <dd>Single-cell reference mapping with Symphony method</dd>

        <dt>

          <code><a href="BBKNN_integrate.html">BBKNN_integrate()</a></code>

        </dt>
        <dd>The BBKNN integration function</dd>

        <dt>

          <code><a href="CSS_integrate.html">CSS_integrate()</a></code>

        </dt>
        <dd>The CSS integration function</dd>

        <dt>

          <code><a href="ComBat_integrate.html">ComBat_integrate()</a></code>

        </dt>
        <dd>The ComBat integration function</dd>

        <dt>

          <code><a href="Conos_integrate.html">Conos_integrate()</a></code>

        </dt>
        <dd>The Conos integration function</dd>

        <dt>

          <code><a href="Harmony_integrate.html">Harmony_integrate()</a></code>

        </dt>
        <dd>The Harmony integration function</dd>

        <dt>

          <code><a href="LIGER_integrate.html">LIGER_integrate()</a></code>

        </dt>
        <dd>The LIGER integration function</dd>

        <dt>

          <code><a href="MNN_integrate.html">MNN_integrate()</a></code>

        </dt>
        <dd>The MNN integration function</dd>

        <dt>

          <code><a href="Scanorama_integrate.html">Scanorama_integrate()</a></code>

        </dt>
        <dd>The Scanorama integration function</dd>

        <dt>

          <code><a href="Seurat_integrate.html">Seurat_integrate()</a></code>

        </dt>
        <dd>The Seurat integration function</dd>

        <dt>

          <code><a href="scVI_integrate.html">scVI_integrate()</a></code>

        </dt>
        <dd>The scVI integration function</dd>

        <dt>

          <code><a href="Uncorrected_integrate.html">Uncorrected_integrate()</a></code>

        </dt>
        <dd>The Uncorrected integration function</dd>

        <dt>

          <code><a href="fastMNN_integrate.html">fastMNN_integrate()</a></code>

        </dt>
        <dd>The fastMNN integration function</dd>
      </dl></div><div class="section level2">

      <h3 id="differential-analysis-and-enrichment">Differential Analysis and Enrichment<a class="anchor" aria-label="anchor" href="#differential-analysis-and-enrichment"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="FindExpressedMarkers.html">FindExpressedMarkers()</a></code>

        </dt>
        <dd>Find Expressed Markers</dd>

        <dt>

          <code><a href="RunDEtest.html">RunDEtest()</a></code>

        </dt>
        <dd>Differential gene test</dd>

        <dt>

          <code><a href="RunGSEA.html">RunGSEA()</a></code>

        </dt>
        <dd>Perform the enrichment analysis (GSEA) on the genes</dd>

        <dt>

          <code><a href="RunEnrichment.html">RunEnrichment()</a></code>

        </dt>
        <dd>Perform the enrichment analysis (over-representation) on the genes</dd>

        <dt>

          <code><a href="RunDynamicEnrichment.html">RunDynamicEnrichment()</a></code>

        </dt>
        <dd>RunDynamicEnrichment</dd>

        <dt>

          <code><a href="RunProportionTest.html">RunProportionTest()</a></code>

        </dt>
        <dd>Proportion Test</dd>
      </dl></div><div class="section level2">

      <h3 id="trajectory-analysis">Trajectory Analysis<a class="anchor" aria-label="anchor" href="#trajectory-analysis"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="RunPAGA.html">RunPAGA()</a></code>

        </dt>
        <dd>Run PAGA analysis</dd>

        <dt>

          <code><a href="RunPalantir.html">RunPalantir()</a></code>

        </dt>
        <dd>Run Palantir analysis</dd>

        <dt>

          <code><a href="RunSlingshot.html">RunSlingshot()</a></code>

        </dt>
        <dd>RunSlingshot</dd>

        <dt>

          <code><a href="RunSCVELO.html">RunSCVELO()</a></code>

        </dt>
        <dd>Run scVelo workflow</dd>

        <dt>

          <code><a href="RunWOT.html">RunWOT()</a></code>

        </dt>
        <dd>Run WOT analysis</dd>

        <dt>

          <code><a href="RunDynamicFeatures.html">RunDynamicFeatures()</a></code>

        </dt>
        <dd>RunDynamicFeatures</dd>
      </dl></div><div class="section level2">

      <h3 id="cell-communication-analysis">Cell Communication Analysis<a class="anchor" aria-label="anchor" href="#cell-communication-analysis"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="RunCellChat.html">RunCellChat()</a></code>

        </dt>
        <dd>Run CellChat</dd>
      </dl></div><div class="section level2">

      <h3 id="visualization-functions">Visualization Functions<a class="anchor" aria-label="anchor" href="#visualization-functions"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="CellChatPlot.html">CellChatPlot()</a></code>

        </dt>
        <dd>Plot CellChat analysis results</dd>

        <dt>

          <code><a href="CellDimPlot.html">CellDimPlot()</a></code>

        </dt>
        <dd>Cell Dimensional Plot</dd>

        <dt>

          <code><a href="CellDimPlot3D.html">CellDimPlot3D()</a></code>

        </dt>
        <dd>3D-Dimensional reduction plot for cell classification visualization.</dd>

        <dt>

          <code><a href="CellDensityPlot.html">CellDensityPlot()</a></code>

        </dt>
        <dd>Cell density plot</dd>

        <dt>

          <code><a href="CellCorHeatmap.html">CellCorHeatmap()</a></code>

        </dt>
        <dd>The Cell Correlation Heatmap</dd>

        <dt>

          <code><a href="CellStatPlot.html">CellStatPlot()</a></code>

        </dt>
        <dd>Statistical plot of cells</dd>

        <dt>

          <code><a href="DynamicPlot.html">DynamicPlot()</a></code>

        </dt>
        <dd>Plot dynamic features across pseudotime</dd>

        <dt>

          <code><a href="DynamicHeatmap.html">DynamicHeatmap()</a></code>

        </dt>
        <dd>Heatmap plot for dynamic features along lineages</dd>

        <dt>

          <code><a href="EnrichmentPlot.html">EnrichmentPlot()</a></code>

        </dt>
        <dd>Enrichment Plot</dd>

        <dt>

          <code><a href="FeatureDimPlot.html">FeatureDimPlot()</a></code>

        </dt>
        <dd>Visualize feature values on a 2-dimensional reduction plot</dd>

        <dt>

          <code><a href="FeatureDimPlot3D.html">FeatureDimPlot3D()</a></code>

        </dt>
        <dd>3D-Dimensional reduction plot for gene expression visualization.</dd>

        <dt>

          <code><a href="FeatureHeatmap.html">FeatureHeatmap()</a></code>

        </dt>
        <dd>Feature Heatmap</dd>

        <dt>

          <code><a href="FeatureCorPlot.html">FeatureCorPlot()</a></code>

        </dt>
        <dd>Features correlation plot</dd>

        <dt>

          <code><a href="FeatureStatPlot.html">FeatureStatPlot()</a></code>

        </dt>
        <dd>Statistical plot of features</dd>

        <dt>

          <code><a href="GraphPlot.html">GraphPlot()</a></code>

        </dt>
        <dd>Graph Plot</dd>

        <dt>

          <code><a href="GroupHeatmap.html">GroupHeatmap()</a></code>

        </dt>
        <dd>The Group Heatmap</dd>

        <dt>

          <code><a href="GSEAPlot.html">GSEAPlot()</a></code>

        </dt>
        <dd>GSEA Plot</dd>

        <dt>

          <code><a href="LineagePlot.html">LineagePlot()</a></code>

        </dt>
        <dd>Lineage Plot</dd>

        <dt>

          <code><a href="PAGAPlot.html">PAGAPlot()</a></code>

        </dt>
        <dd>PAGA plot</dd>

        <dt>

          <code><a href="ProjectionPlot.html">ProjectionPlot()</a></code>

        </dt>
        <dd>Projection Plot</dd>

        <dt>

          <code><a href="ProportionTestPlot.html">ProportionTestPlot()</a></code>

        </dt>
        <dd>Proportion Test Plot</dd>

        <dt>

          <code><a href="StatPlot.html">StatPlot()</a></code>

        </dt>
        <dd>Statistic Plot</dd>

        <dt>

          <code><a href="TACSPlot.html">TACSPlot()</a></code>

        </dt>
        <dd>Transcript-averaged cell scoring (TACS)</dd>

        <dt>

          <code><a href="VelocityPlot.html">VelocityPlot()</a></code>

        </dt>
        <dd>Velocity Plot</dd>

        <dt>

          <code><a href="VolcanoPlot.html">VolcanoPlot()</a></code>

        </dt>
        <dd>Volcano Plot</dd>
      </dl></div><div class="section level2">

      <h3 id="plotting-functions">Plotting Functions<a class="anchor" aria-label="anchor" href="#plotting-functions"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="compute_velocity_on_grid.html">compute_velocity_on_grid()</a></code>

        </dt>
        <dd>Compute velocity on grid</dd>

        <dt>

          <code><a href="cluster_within_group2.html">cluster_within_group2()</a></code>

        </dt>
        <dd>Cluster within group</dd>

        <dt>

          <code><a href="print.scop_logo.html">print(<i>&lt;scop_logo&gt;</i>)</a></code>

        </dt>
        <dd>print scop logo</dd>
      </dl></div><div class="section level2">

      <h3 id="scexplorer">SCExplorer<a class="anchor" aria-label="anchor" href="#scexplorer"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="CreateDataFile.html">CreateDataFile()</a></code>

        </dt>
        <dd>Create data file</dd>

        <dt>

          <code><a href="CreateMetaFile.html">CreateMetaFile()</a></code>

        </dt>
        <dd>Create Meta File</dd>

        <dt>

          <code><a href="FetchH5.html">FetchH5()</a></code>

        </dt>
        <dd>Fetch data from the hdf5 file</dd>

        <dt>

          <code><a href="PrepareSCExplorer.html">PrepareSCExplorer()</a></code>

        </dt>
        <dd>Prepare Seurat objects for the SCExplorer</dd>

        <dt>

          <code><a href="RunSCExplorer.html">RunSCExplorer()</a></code>

        </dt>
        <dd>Run SCExplorer</dd>
      </dl></div><div class="section level2">

      <h3 id="database-operations">Database Operations<a class="anchor" aria-label="anchor" href="#database-operations"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="ListDB.html">ListDB()</a></code>

        </dt>
        <dd>List cached databases</dd>

        <dt>

          <code><a href="PrepareDB.html">PrepareDB()</a></code>

        </dt>
        <dd>Prepare the gene annotation databases</dd>
      </dl></div><div class="section level2">

      <h3 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a></h3>


      <dl></dl></div><div class="section level2">




      <dl><dt>

          <code><a href="ifnb_sub.html">ifnb_sub</a></code>

        </dt>
        <dd>A subsetted version of 'ifnb' datasets</dd>

        <dt>

          <code><a href="panc8_sub.html">panc8_sub</a></code>

        </dt>
        <dd>A subsetted version of human 'panc8' datasets</dd>

        <dt>

          <code><a href="pancreas_sub.html">pancreas_sub</a></code>

        </dt>
        <dd>A subsetted version of mouse 'pancreas' datasets</dd>

        <dt>

          <code><a href="ref_scHCL.html">ref_scHCL</a></code> <code><a href="ref_scHCL.html">ref_scMCA</a></code> <code><a href="ref_scHCL.html">ref_scZCL</a></code>

        </dt>
        <dd>Reference datasets for cell type annotation in single-cell RNA data</dd>

        <dt>

          <code><a href="words_excluded.html">words_excluded</a></code>

        </dt>
        <dd>Excluded words in keyword enrichment analysis and extraction</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xu Meng, Hao Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

