<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>RunDynamicEnrichment — RunDynamicEnrichment • scop</title><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="RunDynamicEnrichment — RunDynamicEnrichment"><meta name="description" content="This function calculates gene-set scores from the specified database (db) for each lineage using the specified scoring method (score_method).
It then treats these scores as expression values and uses them as input to the RunDynamicFeatures function to identify dynamically enriched terms along the lineage."><meta property="og:description" content="This function calculates gene-set scores from the specified database (db) for each lineage using the specified scoring method (score_method).
It then treats these scores as expression values and uses them as input to the RunDynamicFeatures function to identify dynamically enriched terms along the lineage."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scop</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/mengxu98/scop" title="GitHub" aria-label="GitHub repository"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mengxu98" title="MengXu" aria-label="MengXu's GitHub profile"><span class="fa fas fa-home fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>RunDynamicEnrichment</h1>
      <small class="dont-index">Source: <a href="https://github.com/mengxu98/scop/blob/main/R/RunDynamicEnrichment.R" class="external-link"><code>R/RunDynamicEnrichment.R</code></a></small>
      <div class="d-none name"><code>RunDynamicEnrichment.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates gene-set scores from the specified database (<code>db</code>) for each lineage using the specified scoring method (<code>score_method</code>).
It then treats these scores as expression values and uses them as input to the RunDynamicFeatures function to identify dynamically enriched terms along the lineage.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">RunDynamicEnrichment</span><span class="op">(</span></span>
<span>  <span class="va">srt</span>,</span>
<span>  <span class="va">lineages</span>,</span>
<span>  score_method <span class="op">=</span> <span class="st">"AUCell"</span>,</span>
<span>  layer <span class="op">=</span> <span class="st">"data"</span>,</span>
<span>  assay <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_expcells <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  r.sq <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  dev.expl <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  padjust <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  IDtype <span class="op">=</span> <span class="st">"symbol"</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Homo_sapiens"</span>,</span>
<span>  db <span class="op">=</span> <span class="st">"GO_BP"</span>,</span>
<span>  db_update <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  db_version <span class="op">=</span> <span class="st">"latest"</span>,</span>
<span>  convert_species <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  Ensembl_version <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mirror <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  TERM2GENE <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  TERM2NAME <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  minGSSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  maxGSSize <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">11</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-srt">srt<a class="anchor" aria-label="anchor" href="#arg-srt"></a></dt>
<dd><p>A Seurat object containing the results of differential expression analysis (RunDEtest).
If specified, the genes and groups will be extracted from the Seurat object automatically.
If not specified, the <code>geneID</code> and <code>geneID_groups</code> arguments must be provided.</p></dd>


<dt id="arg-lineages">lineages<a class="anchor" aria-label="anchor" href="#arg-lineages"></a></dt>
<dd><p>A character vector specifying the lineages to plot.</p></dd>


<dt id="arg-score-method">score_method<a class="anchor" aria-label="anchor" href="#arg-score-method"></a></dt>
<dd><p>The method to use for scoring.
Can be <code>"Seurat"</code>, <code>"AUCell"</code>, or <code>"UCell"</code>.
Default is <code>"Seurat"</code>.</p></dd>


<dt id="arg-layer">layer<a class="anchor" aria-label="anchor" href="#arg-layer"></a></dt>
<dd><p>Which layer to use.
Default is <code>"counts"</code>.</p></dd>


<dt id="arg-assay">assay<a class="anchor" aria-label="anchor" href="#arg-assay"></a></dt>
<dd><p>Which assay to use.
If <code>NULL</code>, the default assay of the Seurat object will be used.</p></dd>


<dt id="arg-min-expcells">min_expcells<a class="anchor" aria-label="anchor" href="#arg-min-expcells"></a></dt>
<dd><p>The minimum number of expected cells. Default is <code>20</code>.</p></dd>


<dt id="arg-r-sq">r.sq<a class="anchor" aria-label="anchor" href="#arg-r-sq"></a></dt>
<dd><p>The R-squared threshold. Default is <code>0.2</code>.</p></dd>


<dt id="arg-dev-expl">dev.expl<a class="anchor" aria-label="anchor" href="#arg-dev-expl"></a></dt>
<dd><p>The deviance explained threshold. Default is <code>0.2</code>.</p></dd>


<dt id="arg-padjust">padjust<a class="anchor" aria-label="anchor" href="#arg-padjust"></a></dt>
<dd><p>The p-value adjustment threshold. Default is <code>0.05</code>.</p></dd>


<dt id="arg-idtype">IDtype<a class="anchor" aria-label="anchor" href="#arg-idtype"></a></dt>
<dd><p>A character vector specifying the type of gene IDs in the <code>srt</code> object or <code>geneID</code> argument.
This argument is used to convert the gene IDs to a different type if <code>IDtype</code> is different from <code>result_IDtype</code>.</p></dd>


<dt id="arg-species">species<a class="anchor" aria-label="anchor" href="#arg-species"></a></dt>
<dd><p>A character vector specifying the species for which the gene annotation databases should be prepared.
Can be <code>"Homo_sapiens"</code> or <code>"Mus_musculus"</code>.</p></dd>


<dt id="arg-db">db<a class="anchor" aria-label="anchor" href="#arg-db"></a></dt>
<dd><p>A character vector specifying the annotation sources to be included in the gene annotation databases.
Can be one or more of <code>"GO", "GO_BP", "GO_CC", "GO_MF", "KEGG", "WikiPathway", "Reactome", "CORUM", "MP", "DO", "HPO", "PFAM", "CSPA", "Surfaceome", "SPRomeDB", "VerSeDa", "TFLink", "hTFtarget", "TRRUST", "JASPAR", "ENCODE", "MSigDB", "CellTalk", "CellChat", "Chromosome", "GeneType", "Enzyme", "TF"</code>.</p></dd>


<dt id="arg-db-update">db_update<a class="anchor" aria-label="anchor" href="#arg-db-update"></a></dt>
<dd><p>Whether the gene annotation databases should be forcefully updated.
If set to FALSE, the function will attempt to load the cached databases instead.
Default is <code>FALSE</code>.</p></dd>


<dt id="arg-db-version">db_version<a class="anchor" aria-label="anchor" href="#arg-db-version"></a></dt>
<dd><p>A character vector specifying the version of the gene annotation databases to be retrieved.
Default is <code>"latest"</code>.</p></dd>


<dt id="arg-convert-species">convert_species<a class="anchor" aria-label="anchor" href="#arg-convert-species"></a></dt>
<dd><p>Whether to use a species-converted database when the annotation is missing for the specified species.
Default is <code>TRUE</code>.</p></dd>


<dt id="arg-ensembl-version">Ensembl_version<a class="anchor" aria-label="anchor" href="#arg-ensembl-version"></a></dt>
<dd><p>An integer specifying the Ensembl version.
Default is <code>NULL</code>. If <code>NULL</code>, the latest version will be used.</p></dd>


<dt id="arg-mirror">mirror<a class="anchor" aria-label="anchor" href="#arg-mirror"></a></dt>
<dd><p>Specify an Ensembl mirror to connect to. The valid options here are <code>"www"</code>, <code>"uswest"</code>, <code>"useast"</code>, <code>"asia"</code>.</p></dd>


<dt id="arg-term-gene">TERM2GENE<a class="anchor" aria-label="anchor" href="#arg-term-gene"></a></dt>
<dd><p>A data frame specifying the gene-term mapping for a custom database.
The first column should contain the term IDs, and the second column should contain the gene IDs.</p></dd>


<dt id="arg-term-name">TERM2NAME<a class="anchor" aria-label="anchor" href="#arg-term-name"></a></dt>
<dd><p>A data frame specifying the term-name mapping for a custom database.
The first column should contain the term IDs, and the second column should contain the corresponding term names.</p></dd>


<dt id="arg-mingssize">minGSSize<a class="anchor" aria-label="anchor" href="#arg-mingssize"></a></dt>
<dd><p>The minimum size of a gene set to be considered in the enrichment analysis.</p></dd>


<dt id="arg-maxgssize">maxGSSize<a class="anchor" aria-label="anchor" href="#arg-maxgssize"></a></dt>
<dd><p>The maximum size of a gene set to be considered in the enrichment analysis.</p></dd>


<dt id="arg-cores">cores<a class="anchor" aria-label="anchor" href="#arg-cores"></a></dt>
<dd><p>The number of cores to use for parallelization with <a href="https://rdrr.io/pkg/foreach/man/foreach.html" class="external-link">foreach::foreach</a>.
Default is <code>1</code>.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Whether to print the message.
Default is <code>TRUE</code>.</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>Random seed for reproducibility.
Default is <code>11</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="RunDynamicFeatures.html">RunDynamicFeatures</a>, <a href="DynamicHeatmap.html">DynamicHeatmap</a></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pancreas_sub</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="standard_scop.html">standard_scop</a></span><span class="op">(</span><span class="va">pancreas_sub</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="RunSlingshot.html">RunSlingshot</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>  group.by <span class="op">=</span> <span class="st">"SubCellType"</span>,</span></span>
<span class="r-in"><span>  reduction <span class="op">=</span> <span class="st">"UMAP"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="RunDynamicEnrichment-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="RunDynamicFeatures.html">RunDynamicFeatures</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>  lineages <span class="op">=</span> <span class="st">"Lineage1"</span>,</span></span>
<span class="r-in"><span>  n_candidates <span class="op">=</span> <span class="fl">200</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #FFAF00;">⠙</span> [2026-01-13 09:17:04] Running for <span style="color: #0000BB;">1 </span>[<span style="color: #0000BB;">1</span>/<span style="color: #0000BB;">199</span>] <span style="color: #00BB00;">■                               </span> …</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #FFAF00;">⠹</span> [2026-01-13 09:17:04] Running for <span style="color: #0000BB;">56 </span>[<span style="color: #0000BB;">56</span>/<span style="color: #0000BB;">199</span>] <span style="color: #00BB00;">■■■■■■■■■                      </span>…</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #FFAF00;">⠸</span> [2026-01-13 09:17:04] Running for <span style="color: #0000BB;">142 </span>[<span style="color: #0000BB;">142</span>/<span style="color: #0000BB;">199</span>] <span style="color: #00BB00;">■■■■■■■■■■■■■■■■■■■■■■       </span>…</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> [2026-01-13 09:17:04] Completed <span style="color: #0000BB;">199</span> tasks in <span style="color: #0000BB;">6.7s</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span><span class="va">ht1</span> <span class="op">&lt;-</span> <span class="fu"><a href="DynamicHeatmap.html">DynamicHeatmap</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>  lineages <span class="op">=</span> <span class="st">"Lineage1"</span>,</span></span>
<span class="r-in"><span>  cell_annotation <span class="op">=</span> <span class="st">"SubCellType"</span>,</span></span>
<span class="r-in"><span>  n_split <span class="op">=</span> <span class="fl">4</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'magick' package is suggested to install to give better rasterization.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Set `ht_opt$message = FALSE` to turn off this message.</span>
<span class="r-plt img"><img src="RunDynamicEnrichment-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">ht1</span><span class="op">$</span><span class="va">plot</span></span></span>
<span class="r-plt img"><img src="RunDynamicEnrichment-3.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">pancreas_sub</span> <span class="op">&lt;-</span> <span class="fu">RunDynamicEnrichment</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>  lineages <span class="op">=</span> <span class="st">"Lineage1"</span>,</span></span>
<span class="r-in"><span>  score_method <span class="op">=</span> <span class="st">"UCell"</span>,</span></span>
<span class="r-in"><span>  db <span class="op">=</span> <span class="st">"GO_BP"</span>,</span></span>
<span class="r-in"><span>  species <span class="op">=</span> <span class="st">"Mus_musculus"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in loadNamespace(x):</span> there is no package called ‘R.cache’</span>
<span class="r-in"><span><span class="va">ht2</span> <span class="op">&lt;-</span> <span class="fu"><a href="DynamicHeatmap.html">DynamicHeatmap</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">pancreas_sub</span>,</span></span>
<span class="r-in"><span>  assay <span class="op">=</span> <span class="st">"GO_BP"</span>,</span></span>
<span class="r-in"><span>  lineages <span class="op">=</span> <span class="st">"Lineage1_GO_BP"</span>,</span></span>
<span class="r-in"><span>  cell_annotation <span class="op">=</span> <span class="st">"SubCellType"</span>,</span></span>
<span class="r-in"><span>  n_split <span class="op">=</span> <span class="fl">4</span>,</span></span>
<span class="r-in"><span>  split_method <span class="op">=</span> <span class="st">"kmeans-peaktime"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in DynamicHeatmap(pancreas_sub, assay = "GO_BP", lineages = "Lineage1_GO_BP",     cell_annotation = "SubCellType", n_split = 4, split_method = "kmeans-peaktime"):</span> Lineages: Lineage1_GO_BP is not in the meta data of <span style="color: #0000BB;">&lt;Seurat&gt;</span></span>
<span class="r-in"><span><span class="va">ht2</span><span class="op">$</span><span class="va">plot</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'ht2' not found</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xu Meng, Hao Zhang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

