#' @title Prepare the gene annotation databases
#'
#' @description
#' This function prepares the gene annotation databases for a given species and set of annotation sources.
#' It retrieves the necessary information from various annotation packages or external resources and organizes it into a list.
#' The list contains the annotation data for each specified annotation source.
#'
#' @md
#' @inheritParams GeneConvert
#' @inheritParams thisutils::log_message
#' @param species A character vector specifying the species for which the gene annotation databases should be prepared.
#' Default is `c("Homo_sapiens", "Mus_musculus")`.
#' @param db A character vector specifying the annotation sources to be included in the gene annotation databases.
#' Default is `c("GO", "GO_BP", "GO_CC", "GO_MF", "KEGG", "WikiPathway", "Reactome",
#' "CORUM", "MP", "DO", "HPO", "PFAM", "CSPA", "Surfaceome", "SPRomeDB", "VerSeDa",
#' "TFLink", "hTFtarget", "TRRUST", "JASPAR", "ENCODE", "MSigDB",
#' "CellTalk", "CellChat", "Chromosome", "GeneType", "Enzyme", "TF")`.
#' @param db_IDtypes A character vector specifying the desired ID types to be used for gene identifiers in the gene annotation databases.
#' Default is `c("symbol", "entrez_id", "ensembl_id")`.
#' @param db_version A character vector specifying the version of the gene annotation databases to be retrieved.
#' Default is `"latest"`.
#' @param db_update Whether the gene annotation databases should be forcefully updated.
#' If set to FALSE, the function will attempt to load the cached databases instead.
#' Default is `FALSE`.
#' @param convert_species Whether to use a species-converted database when the annotation is missing for the specified species.
#' The default value is `TRUE`.
#' @param custom_TERM2GENE A data frame containing a custom TERM2GENE mapping for the specified species and annotation source.
#' Default is `NULL`.
#' @param custom_TERM2NAME A data frame containing a custom TERM2NAME mapping for the specified species and annotation source.
#' Default is `NULL`.
#' @param custom_species A character vector specifying the species name to be used in a custom database.
#' Default is `NULL`.
#' @param custom_IDtype A character vector specifying the ID type to be used in a custom database.
#' Default is `NULL`.
#' @param custom_version A character vector specifying the version to be used in a custom database.
#' Default is `NULL`.
#'
#' @details
#' The `PrepareDB` function prepares gene annotation databases for a given species and set of annotation sources.
#' It retrieves the necessary information from various annotation packages or external resources and organizes it into a list.
#' The function also supports creating custom databases based on user-provided gene sets.
#'
#' @return A list containing the prepared gene annotation databases:
#'   \itemize{
#'     \item \code{TERM2GENE:} mapping of gene identifiers to terms.
#'     \item \code{TERM2NAME:} mapping of terms to their names.
#'     \item \code{semData:} semantic similarity data for gene sets (only for Gene Ontology terms).
#'     }
#'
#' @seealso [ListDB]
#'
#' @export
#'
#' @examples
#' \dontrun{
#' db_list <- PrepareDB(
#'   species = "Homo_sapiens",
#'   db = "GO_BP"
#' )
#' ListDB(
#'   species = "Homo_sapiens",
#'   db = "GO_BP"
#' )
#' head(
#'   db_list[["Homo_sapiens"]][["GO_BP"]][["TERM2GENE"]]
#' )
#'
#' # Based on homologous gene conversion,
#' # prepare a gene annotation database that originally does not exist in the species.
#' db_list <- PrepareDB(
#'   species = "Homo_sapiens",
#'   db = "MP"
#' )
#' ListDB(
#'   species = "Homo_sapiens",
#'   db = "MP"
#' )
#' head(
#'   db_list[["Homo_sapiens"]][["MP"]][["TERM2GENE"]]
#' )
#'
#' # Prepare databases for other species
#' db_list <- PrepareDB(
#'   species = "Macaca_fascicularis",
#'   db = "GO_BP"
#' )
#' ListDB(
#'   species = "Macaca_fascicularis",
#'   db = "GO_BP"
#' )
#' head(
#'   db_list[["Macaca_fascicularis"]][["GO_BP"]][["TERM2GENE"]]
#' )
#'
#' db_list <- PrepareDB(
#'   species = "Saccharomyces_cerevisiae",
#'   db = "GO_BP"
#' )
#' ListDB(
#'   species = "Saccharomyces_cerevisiae",
#'   db = "GO_BP"
#' )
#' head(
#'   db_list[["Saccharomyces_cerevisiae"]][["GO_BP"]][["TERM2GENE"]]
#' )
#'
#' # Prepare databases for Arabidopsis (plant)
#' db_list <- PrepareDB(
#'   species = "Arabidopsis_thaliana",
#'   db = c(
#'     "GO_BP", "GO_CC", "GO_MF", "KEGG", "WikiPathway",
#'     "ENZYME", "Chromosome"
#'   ),
#'   biomart = "plants_mart"
#' )
#' head(
#'   db_list[["Arabidopsis_thaliana"]][["KEGG"]][["TERM2GENE"]]
#' )
#'
#' # You can also build a custom database based on the gene sets you have
#' ccgenes <- CycGenePrefetch("Homo_sapiens")
#' custom_TERM2GENE <- rbind(
#'   data.frame(
#'     term = "S_genes",
#'     gene = ccgenes[["cc_S_genes"]]
#'   ),
#'   data.frame(
#'     term = "G2M_genes",
#'     gene = ccgenes[["cc_G2M_genes"]]
#'   )
#' )
#' str(custom_TERM2GENE)
#'
#' # Set convert_species = TRUE to build a custom database for both species,
#' # with the name "CellCycle"
#' db_list <- PrepareDB(
#'   species = c("Homo_sapiens", "Mus_musculus"),
#'   db = "CellCycle",
#'   convert_species = TRUE,
#'   custom_TERM2GENE = custom_TERM2GENE,
#'   custom_species = "Homo_sapiens",
#'   custom_IDtype = "symbol",
#'   custom_version = "Seurat_v5"
#' )
#' ListDB(db = "CellCycle")
#'
#' db_list <- PrepareDB(species = "Mus_musculus", db = "CellCycle")
#' head(
#'   db_list[["Mus_musculus"]][["CellCycle"]][["TERM2GENE"]]
#' )
#' }
PrepareDB <- function(
    species = c("Homo_sapiens", "Mus_musculus"),
    db = c(
      "GO",
      "GO_BP",
      "GO_CC",
      "GO_MF",
      "KEGG",
      "WikiPathway",
      "Reactome",
      "CORUM",
      "MP",
      "DO",
      "HPO",
      "PFAM",
      "CSPA",
      "Surfaceome",
      "SPRomeDB",
      "VerSeDa",
      "TFLink",
      "hTFtarget",
      "TRRUST",
      "JASPAR",
      "ENCODE",
      "MSigDB",
      "CellTalk",
      "CellChat",
      "Chromosome",
      "GeneType",
      "Enzyme",
      "TF"
    ),
    db_IDtypes = c("symbol", "entrez_id", "ensembl_id"),
    db_version = "latest",
    db_update = FALSE,
    convert_species = TRUE,
    Ensembl_version = 103,
    mirror = NULL,
    biomart = NULL,
    max_tries = 5,
    custom_TERM2GENE = NULL,
    custom_TERM2NAME = NULL,
    custom_species = NULL,
    custom_IDtype = NULL,
    custom_version = NULL,
    verbose = TRUE) {
  db_list <- list()
  for (sps in species) {
    log_message(
      "Species: ", sps,
      verbose = verbose
    )
    default_id_types <- list(
      "GO" = "entrez_id",
      "GO_BP" = "entrez_id",
      "GO_CC" = "entrez_id",
      "GO_MF" = "entrez_id",
      "KEGG" = "entrez_id",
      "WikiPathway" = "entrez_id",
      "Reactome" = "entrez_id",
      "CORUM" = "symbol",
      "MP" = "symbol",
      "DO" = "symbol",
      "HPO" = "symbol",
      "PFAM" = "entrez_id",
      "Chromosome" = "entrez_id",
      "GeneType" = "entrez_id",
      "Enzyme" = "entrez_id",
      "TF" = "symbol",
      "CSPA" = "symbol",
      "Surfaceome" = "symbol",
      "SPRomeDB" = "entrez_id",
      "VerSeDa" = "symbol",
      "TFLink" = "symbol",
      "hTFtarget" = "symbol",
      "TRRUST" = "symbol",
      "JASPAR" = "symbol",
      "ENCODE" = "symbol",
      "MSigDB" = c("symbol", "ensembl_id"),
      "CellTalk" = "symbol",
      "CellChat" = "symbol"
    )
    if (!is.null(custom_TERM2GENE)) {
      if (length(db) > 1) {
        log_message(
          "When building a custom database, the length of {.arg db} must be 1",
          message_type = "error"
        )
      }
      if (is.null(custom_IDtype) || is.null(custom_species) || is.null(custom_version)) {
        log_message(
          "When building a custom database, {.arg custom_IDtype}, {.arg custom_species} and {.arg custom_version} must be provided",
          message_type = "error"
        )
      }
      custom_IDtype <- match.arg(
        custom_IDtype,
        choices = c("symbol", "entrez_id", "ensembl_id")
      )
      default_id_types[[db]] <- custom_IDtype
    }

    if (isFALSE(db_update) && is.null(custom_TERM2GENE)) {
      for (term in db) {
        dbinfo <- ListDB(species = sps, db = term)
        if (nrow(dbinfo) > 0 && !is.null(dbinfo)) {
          if (db_version == "latest") {
            pathname <- dbinfo[
              order(dbinfo[["timestamp"]], decreasing = TRUE)[1],
              "file"
            ]
          } else {
            pathname <- dbinfo[
              grep(db_version, dbinfo[["db_version"]], fixed = TRUE)[1],
              "file"
            ]
            if (is.na(pathname)) {
              log_message(
                "There is no ",
                db_version,
                " version of the database. Use the latest version",
                message_type = "warning"
              )
              pathname <- dbinfo[
                order(dbinfo[["timestamp"]], decreasing = TRUE)[1],
                "file"
              ]
            }
          }
          if (!is.na(pathname)) {
            header <- R.cache::readCacheHeader(pathname)
            cached_version <- strsplit(header[["comment"]], "\\|")[[1]][1]
            timestamp <- format(header[["timestamp"]], "%Y-%m-%d %H:%M:%S")
            log_message(
              "Loading cached: {.pkg {term}} version: {.pkg {cached_version}} created: {.val {timestamp}}"
            )
            db_loaded <- R.cache::loadCache(pathname = pathname)
            Sys.sleep(0.5)
            db_list[[sps]][[term]] <- db_loaded
          }
        }
      }
    }

    db_species <- stats::setNames(object = rep(sps, length(db)), nm = db)

    sp <- unlist(strsplit(sps, split = "_"))
    org_sp <- paste0(
      "org.",
      paste0(substring(sp, 1, 1), collapse = ""),
      ".eg.db"
    )
    org_key <- "ENTREZID"
    if (sps == "Arabidopsis_thaliana") {
      biomart <- "plants_mart"
      org_sp <- "org.At.tair.db"
      org_key <- "TAIR"
      default_id_types[c(
        "GO",
        "GO_BP",
        "GO_CC",
        "GO_MF",
        "PFAM",
        "Chromosome",
        "GeneType",
        "Enzyme"
      )] <- "tair_locus"
    }
    if (sps == "Saccharomyces_cerevisiae") {
      org_sp <- "org.Sc.sgd.db"
      # org_key <- "SGD"
      # default_id_types[c("GO", "GO_BP", "GO_CC", "GO_MF", "PFAM", "Chromosome", "GeneType", "Enzyme")] <- "sgd_gene"
    }

    ## Prepare -----------------
    if (any(!sps %in% names(db_list)) || any(!db %in% names(db_list[[sps]]))) {
      orgdb_dependent <- c(
        "GO",
        "GO_BP",
        "GO_CC",
        "GO_MF",
        "PFAM",
        "Chromosome",
        "GeneType",
        "Enzyme"
      )
      if (any(orgdb_dependent %in% db)) {
        status <- tryCatch(
          {
            check_r(c(org_sp, "GO.db", "GOSemSim"))
          },
          error = identity
        )
        if (inherits(status, "error")) {
          log_message(
            "Annotation package {.pkg {org_sp}} does not exist",
            message_type = "warning"
          )
          if (isTRUE(convert_species)) {
            db_to_convert <- intersect(db, orgdb_dependent)
            log_message(
              "Use the human annotation to create the {.pkg {db_to_convert}} database for {.val {sps}}",
              message_type = "warning"
            )
            org_sp <- "org.Hs.eg.db"
            db_species[db_to_convert] <- "Homo_sapiens"
          } else {
            log_message(
              "Stop the preparation",
              message_type = "error"
            )
          }
        }
        suppressPackageStartupMessages(require(
          org_sp,
          character.only = TRUE,
          quietly = TRUE
        ))
        orgdb <- get(org_sp)
      }
      if ("PFAM" %in% db) {
        check_r("PFAM.db")
      }
      if ("Reactome" %in% db) {
        check_r("reactome.db")
      }

      if (is.null(custom_TERM2GENE)) {
        ## GO -----------------------
        go_categories <- c("GO", "GO_BP", "GO_CC", "GO_MF")
        if (any(db %in% go_categories) &&
          any(!intersect(db, go_categories) %in% names(db_list[[sps]]))
        ) {
          terms <- db[db %in% go_categories]
          bg <- suppressMessages(
            AnnotationDbi::select(
              orgdb,
              keys = AnnotationDbi::keys(orgdb),
              columns = c("GOALL", org_key)
            )
          )
          bg <- unique(bg[
            !is.na(bg[["GOALL"]]),
            c("GOALL", "ONTOLOGYALL", org_key),
            drop = FALSE
          ])
          bg2 <- suppressMessages(
            AnnotationDbi::select(
              GO.db::GO.db,
              keys = AnnotationDbi::keys(GO.db::GO.db),
              columns = c("GOID", "TERM")
            )
          )
          bg <- merge(
            x = bg,
            by.x = "GOALL",
            y = bg2,
            by.y = "GOID",
            all.x = TRUE
          )
          for (subterm in terms) {
            log_message("Preparing database: {.pkg {subterm}}")
            if (subterm == "GO") {
              TERM2GENE <- bg[, c("GOALL", org_key)]
              TERM2NAME <- bg[, c("GOALL", "TERM")]
              colnames(TERM2GENE) <- c("Term", default_id_types[[subterm]])
              colnames(TERM2NAME) <- c("Term", "Name")
              TERM2NAME[["ONTOLOGY"]] <- bg[["ONTOLOGYALL"]]
              semData <- NULL
            } else {
              simpleterm <- unlist(strsplit(subterm, split = "_"))[2]
              TERM2GENE <- bg[
                which(bg[["ONTOLOGYALL"]] %in% simpleterm),
                c("GOALL", org_key)
              ]
              TERM2NAME <- bg[
                which(bg[["ONTOLOGYALL"]] %in% simpleterm),
                c("GOALL", "TERM")
              ]
              colnames(TERM2GENE) <- c("Term", default_id_types[[subterm]])
              colnames(TERM2NAME) <- c("Term", "Name")
              TERM2NAME[["ONTOLOGY"]] <- simpleterm
              semData <- suppressMessages(
                GOSemSim::godata(
                  annoDb = orgdb,
                  ont = simpleterm
                )
              )
            }
            TERM2GENE <- stats::na.omit(unique(TERM2GENE))
            TERM2NAME <- stats::na.omit(unique(TERM2NAME))
            version <- utils::packageVersion(org_sp)
            db_list[[db_species[subterm]]][[subterm]][[
              "TERM2GENE"
            ]] <- TERM2GENE
            db_list[[db_species[subterm]]][[subterm]][[
              "TERM2NAME"
            ]] <- TERM2NAME
            db_list[[db_species[subterm]]][[subterm]][["semData"]] <- semData
            db_list[[db_species[subterm]]][[subterm]][["version"]] <- version
            if (sps == db_species[subterm]) {
              R.cache::saveCache(
                db_list[[db_species[subterm]]][[subterm]],
                key = list(version, as.character(db_species[subterm]), subterm),
                comment = paste0(
                  version,
                  " nterm:",
                  length(TERM2NAME[[1]]),
                  "|",
                  db_species[subterm],
                  "-",
                  subterm
                )
              )
            }
          }
        }

        ## KEGG -----------------
        if (any(db == "KEGG") && (!"KEGG" %in% names(db_list[[sps]]))) {
          log_message("Preparing database: KEGG")
          check_r("httr")
          orgs <- kegg_get("https://rest.kegg.jp/list/organism")
          kegg_sp <- orgs[
            grep(gsub(pattern = "_", replacement = " ", x = sps), orgs[, 3]),
            2
          ]
          if (length(kegg_sp) == 0) {
            db_species_name <- db_species["KEGG"]
            log_message(
              "Failed to prepare the KEGG database for {.val {db_species_name}}",
              message_type = "warning"
            )
            if (isTRUE(convert_species) && db_species_name != "Homo_sapiens") {
              log_message(
                "Use the human annotation to create the KEGG database for {.val {sps}}",
                message_type = "warning"
              )
              db_species["KEGG"] <- "Homo_sapiens"
              kegg_sp <- "hsa"
              return(NULL)
            } else {
              log_message(
                "Stop the preparation",
                message_type = "error"
              )
            }
          }
          kegg_db <- "pathway"

          kegg_pathwaygene_url <- paste0(
            "https://rest.kegg.jp/link/",
            kegg_sp,
            "/",
            kegg_db,
            collapse = ""
          )
          TERM2GENE <- kegg_get(kegg_pathwaygene_url)
          colnames(TERM2GENE) <- c("Pathway", "KEGG_ID")
          kegg_geneconversion_url <- paste0(
            "https://rest.kegg.jp/conv/ncbi-geneid/",
            kegg_sp
          )
          GENECONV <- kegg_get(kegg_geneconversion_url)
          colnames(GENECONV) <- c("KEGG_ID", "ENTREZID")
          TERM2GENE <- merge(
            x = TERM2GENE,
            y = GENECONV,
            by = "KEGG_ID",
            all.x = TRUE
          )
          TERM2GENE[, "Pathway"] <- gsub(
            pattern = "[^:]+:",
            replacement = "",
            x = TERM2GENE[, "Pathway"]
          )
          TERM2GENE[, "ENTREZID"] <- gsub(
            pattern = "[^:]+:",
            replacement = "",
            x = TERM2GENE[, "ENTREZID"]
          )
          TERM2GENE <- TERM2GENE[, c("Pathway", "ENTREZID")]

          kegg_pathwayname_url <- paste0(
            "https://rest.kegg.jp/list/",
            kegg_db,
            "/",
            kegg_sp,
            collapse = ""
          )
          TERM2NAME <- kegg_get(kegg_pathwayname_url)
          colnames(TERM2NAME) <- c("Pathway", "Name")
          TERM2NAME[, "Pathway"] <- gsub(
            pattern = "[^:]+:",
            replacement = "",
            x = TERM2NAME[, "Pathway"]
          )
          TERM2NAME[, "Name"] <- gsub(
            pattern = paste0(
              " - ",
              paste0(
                unlist(strsplit(db_species["KEGG"], split = "_")),
                collapse = " "
              ),
              ".*$"
            ),
            replacement = "",
            x = TERM2NAME[, "Name"]
          )
          TERM2NAME <- TERM2NAME[
            TERM2NAME[, "Pathway"] %in% TERM2GENE[, "Pathway"], ,
            drop = FALSE
          ]

          colnames(TERM2GENE) <- c("Term", default_id_types[["KEGG"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          # kegg_info <- readLines("https://rest.kegg.jp/info/hsa")
          kegg_info <- strsplit(
            httr::content(httr::GET(paste0(
              "https://rest.kegg.jp/info/",
              kegg_sp
            ))),
            split = "\n"
          )[[1]]
          version <- gsub(
            ".*(?=Release)",
            replacement = "",
            x = kegg_info[grepl("Release", x = kegg_info)],
            perl = TRUE
          )
          db_list[[db_species["KEGG"]]][["KEGG"]][["TERM2GENE"]] <- TERM2GENE
          db_list[[db_species["KEGG"]]][["KEGG"]][["TERM2NAME"]] <- TERM2NAME
          db_list[[db_species["KEGG"]]][["KEGG"]][["version"]] <- version
          if (sps == db_species["KEGG"]) {
            R.cache::saveCache(
              db_list[[db_species["KEGG"]]][["KEGG"]],
              key = list(version, as.character(db_species["KEGG"]), "KEGG"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["KEGG"],
                "-KEGG"
              )
            )
          }
        }

        ## WikiPathway -----------------
        if (
          any(db == "WikiPathway") &&
            (!"WikiPathway" %in% names(db_list[[sps]]))
        ) {
          log_message("Preparing database: WikiPathway")
          tempdir <- tempdir()
          gmt_files <- list.files(tempdir)[grep(
            ".gmt",
            x = list.files(tempdir)
          )]
          if (length(gmt_files) > 0) {
            file.remove(paste0(tempdir, "/", gmt_files))
          }
          temp <- tempfile()
          download(
            url = "https://wikipathways-data.wmcloud.org/current/gmt",
            destfile = temp
          )
          lines <- paste0(readLines(temp, warn = FALSE), collapse = " ")
          gmtfiles <- unlist(regmatches(
            lines,
            m = gregexpr("(?<=>)wikipathways-\\S+\\.gmt\\b", lines, perl = TRUE)
          ))
          wiki_sp <- sps
          gmtfile <- gmtfiles[grep(wiki_sp, gmtfiles, fixed = TRUE)]
          if (length(gmtfile) == 0) {
            db_species_name <- db_species["WikiPathway"]
            log_message(
              "Failed to prepare the WikiPathway database for {.val {db_species_name}}",
              message_type = "warning"
            )
            if (isTRUE(convert_species) && db_species_name != "Homo_sapiens") {
              log_message(
                "Use the human annotation to create the WikiPathway database for {.val {sps}}",
                message_type = "warning"
              )
              db_species["WikiPathway"] <- "Homo_sapiens"
              wiki_sp <- "Homo_sapiens"
              gmtfile <- gmtfiles[grep(wiki_sp, gmtfiles, fixed = TRUE)]
            } else {
              log_message(
                "Stop the preparation",
                message_type = "error"
              )
            }
          }
          version <- strsplit(gmtfile, split = "-")[[1]][[2]]
          download(
            url = paste0(
              "https://wikipathways-data.wmcloud.org/current/gmt/",
              gmtfile
            ),
            destfile = temp
          )
          wiki_gmt <- clusterProfiler::read.gmt(temp)
          unlink(temp)
          wiki_gmt <- apply(wiki_gmt, 1, function(x) {
            wikiid <- strsplit(x[["term"]], split = "%")[[1]][3]
            wikiterm <- strsplit(x[["term"]], split = "%")[[1]][1]
            gmt <- x[["gene"]]
            data.frame(
              v0 = wikiid,
              v1 = gmt,
              v2 = wikiterm,
              stringsAsFactors = FALSE
            )
          })
          bg <- do.call(rbind.data.frame, wiki_gmt)
          TERM2GENE <- bg[, c(1, 2)]
          TERM2NAME <- bg[, c(1, 3)]
          colnames(TERM2GENE) <- c("Term", default_id_types[["WikiPathway"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["WikiPathway"]]][["WikiPathway"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["WikiPathway"]]][["WikiPathway"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["WikiPathway"]]][["WikiPathway"]][[
            "version"
          ]] <- version
          if (sps == db_species["WikiPathway"]) {
            R.cache::saveCache(
              db_list[[db_species["WikiPathway"]]][["WikiPathway"]],
              key = list(
                version,
                as.character(db_species["WikiPathway"]),
                "WikiPathway"
              ),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["WikiPathway"],
                "-WikiPathway"
              )
            )
          }
        }

        ## Reactome -----------------
        if (any(db == "Reactome") && (!"Reactome" %in% names(db_list[[sps]]))) {
          log_message("Preparing database: {.pkg Reactome}")
          reactome_sp <- gsub(pattern = "_", replacement = " ", x = sps)
          df_all <- suppressMessages(
            AnnotationDbi::select(
              reactome.db::reactome.db,
              keys = AnnotationDbi::keys(reactome.db::reactome.db),
              columns = c("PATHID", "PATHNAME")
            )
          )
          df <- df_all[
            grepl(
              pattern = paste0("^", reactome_sp, ": "),
              x = df_all$PATHNAME
            ), ,
            drop = FALSE
          ]
          if (nrow(df) == 0) {
            if (
              isTRUE(convert_species) &&
                db_species["Reactome"] != "Homo_sapiens"
            ) {
              log_message(
                "Use the human annotation to create the Reactome database for {.val {sps}}",
                message_type = "warning"
              )
              db_species["Reactome"] <- "Homo_sapiens"
              reactome_sp <- gsub(
                pattern = "_",
                replacement = " ",
                x = "Homo_sapiens"
              )
              df <- df_all[
                grepl(
                  pattern = paste0("^", reactome_sp, ": "),
                  x = df_all$PATHNAME
                ), ,
                drop = FALSE
              ]
            } else {
              log_message(
                "Stop the preparation",
                message_type = "error"
              )
            }
          }
          df <- stats::na.omit(df)
          df$PATHNAME <- gsub(
            x = df$PATHNAME,
            pattern = paste0("^", reactome_sp, ": "),
            replacement = "",
            perl = TRUE
          )
          TERM2GENE <- df[, c(2, 1)]
          TERM2NAME <- df[, c(2, 3)]
          colnames(TERM2GENE) <- c("Term", default_id_types[["Reactome"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          version <- utils::packageVersion("reactome.db")
          db_list[[db_species["Reactome"]]][["Reactome"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["Reactome"]]][["Reactome"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["Reactome"]]][["Reactome"]][[
            "version"
          ]] <- version
          if (sps == db_species["Reactome"]) {
            R.cache::saveCache(
              db_list[[db_species["Reactome"]]][["Reactome"]],
              key = list(
                version,
                as.character(db_species["Reactome"]),
                "Reactome"
              ),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["Reactome"],
                "-Reactome"
              )
            )
          }
        }

        ## CORUM -----------------
        if (any(db == "CORUM") && (!"CORUM" %in% names(db_list[[sps]]))) {
          if (!sps %in% c("Homo_sapiens")) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the CORUM database for {.val {sps}}",
                message_type = "warning"
              )
              db_species["CORUM"] <- "Homo_sapiens"
            } else {
              log_message(
                "CORUM database only support Homo_sapiens. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: corum")
          url <- "https://maayanlab.cloud/static/hdfs/harmonizome/data/corum/gene_set_library_crisp.gmt.gz"
          temp <- tempfile(fileext = ".gz")
          download(url = url, destfile = temp)
          R.utils::gunzip(temp)
          TERM2GENE <- clusterProfiler::read.gmt(gsub(".gz", "", temp))
          version <- "Harmonizome 3.0"
          TERM2NAME <- TERM2GENE[, c(1, 1)]
          colnames(TERM2GENE) <- c("Term", default_id_types[["CORUM"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["CORUM"]]][["CORUM"]][["TERM2GENE"]] <- TERM2GENE
          db_list[[db_species["CORUM"]]][["CORUM"]][["TERM2NAME"]] <- TERM2NAME
          db_list[[db_species["CORUM"]]][["CORUM"]][["version"]] <- version
          if (sps == db_species["CORUM"]) {
            R.cache::saveCache(
              db_list[[db_species["CORUM"]]][["CORUM"]],
              key = list(version, as.character(db_species["CORUM"]), "CORUM"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["CORUM"],
                "-CORUM"
              )
            )
          }
        }

        ## MP -----------------
        if (any(db == "MP") && (!"MP" %in% names(db_list[[sps]]))) {
          if (sps != "Mus_musculus") {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the mouse annotation to create the MP database for ",
                sps,
                message_type = "warning"
              )
              db_species["MP"] <- "Mus_musculus"
            } else {
              log_message(
                "MP database only support Mus_musculus. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: MP")
          temp <- tempfile()
          download(
            url = "http://www.informatics.jax.org/downloads/reports/",
            destfile = temp
          )
          version <- readLines(temp, warn = FALSE)
          version <- version[grep("MGI_PhenoGenoMP.rpt", version)]
          version <- strsplit(version, split = "  </td><td align=\"right\">")[[
            1
          ]][2]
          download(
            url = "http://www.informatics.jax.org/downloads/reports/VOC_MammalianPhenotype.rpt",
            destfile = temp
          )
          mp_name <- utils::read.table(
            temp,
            header = FALSE,
            sep = "\t",
            fill = TRUE,
            quote = ""
          )
          rownames(mp_name) <- mp_name[, 1]
          download(
            url = "http://www.informatics.jax.org/downloads/reports/MGI_Gene_Model_Coord.rpt",
            destfile = temp
          )
          gene_id <- utils::read.table(
            temp,
            header = FALSE,
            row.names = NULL,
            sep = "\t",
            fill = TRUE,
            quote = ""
          )
          gene_id <- gene_id[, 1:15]
          colnames(gene_id) <- gene_id[1, ]
          gene_id <- gene_id[
            gene_id[, 2] %in% c("Gene", "Pseudogene"), ,
            drop = FALSE
          ]
          rownames(gene_id) <- gene_id[, 1]

          download(
            url = "http://www.informatics.jax.org/downloads/reports/MGI_GenePheno.rpt",
            destfile = temp
          ) # 32.4 MB
          mp_gene <- utils::read.table(
            temp,
            header = FALSE,
            sep = "\t",
            fill = TRUE,
            quote = ""
          )
          mp_gene[["symbol"]] <- gene_id[mp_gene[["V7"]], "3. marker symbol"]
          mp_gene[["MP"]] <- mp_name[mp_gene[, "V5"], 2]
          TERM2GENE <- mp_gene[, c("V5", "symbol")]
          TERM2NAME <- mp_gene[, c("V5", "MP")]

          colnames(TERM2GENE) <- c("Term", default_id_types[["MP"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["MP"]]][["MP"]][["TERM2GENE"]] <- TERM2GENE
          db_list[[db_species["MP"]]][["MP"]][["TERM2NAME"]] <- TERM2NAME
          db_list[[db_species["MP"]]][["MP"]][["version"]] <- version
          if (sps == db_species["MP"]) {
            R.cache::saveCache(
              db_list[[db_species["MP"]]][["MP"]],
              key = list(version, as.character(db_species["MP"]), "MP"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["MP"],
                "-MP"
              )
            )
          }
        }

        ## DO -----------------
        if (any(db == "DO") && (!"DO" %in% names(db_list[[sps]]))) {
          log_message("Preparing database: DO")
          temp <- tempfile(fileext = ".tsv.gz")
          download(
            url = "https://fms.alliancegenome.org/download/DISEASE-ALLIANCE_COMBINED.tsv.gz",
            destfile = temp
          )
          R.utils::gunzip(temp)
          do_all <- utils::read.table(
            gsub(".gz", "", temp),
            header = TRUE,
            sep = "\t",
            fill = TRUE,
            quote = ""
          )
          version <- gsub(
            pattern = ".*Alliance Database Version: ",
            replacement = "",
            x = grep(
              "Alliance Database Version",
              readLines(gsub(".gz", "", temp), warn = FALSE),
              perl = TRUE,
              value = TRUE
            )
          )
          unlink(temp)
          do_sp <- gsub(pattern = "_", replacement = " ", x = sps)
          do_df <- do_all[
            do_all[["DBobjectType"]] == "gene" &
              do_all[["SpeciesName"]] == do_sp, ,
            drop = FALSE
          ]
          if (nrow(do_df) == 0) {
            if (isTRUE(convert_species) && db_species["DO"] != "Homo_sapiens") {
              log_message(
                "Use the human annotation to create the DO database for ",
                sps,
                message_type = "warning"
              )
              db_species["DO"] <- "Homo_sapiens"
              do_sp <- gsub(
                pattern = "_",
                replacement = " ",
                x = "Homo_sapiens"
              )
              do_df <- do_all[
                do_all[["DBobjectType"]] == "gene" &
                  do_all[["SpeciesName"]] == do_sp, ,
                drop = FALSE
              ]
            } else {
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          TERM2GENE <- do_df[, c("DOID", "DBObjectSymbol")]
          TERM2NAME <- do_df[, c("DOID", "DOtermName")]
          colnames(TERM2GENE) <- c("Term", default_id_types[["DO"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["DO"]]][["DO"]][["TERM2GENE"]] <- TERM2GENE
          db_list[[db_species["DO"]]][["DO"]][["TERM2NAME"]] <- TERM2NAME
          db_list[[db_species["DO"]]][["DO"]][["version"]] <- version
          if (sps == db_species["DO"]) {
            R.cache::saveCache(
              db_list[[db_species["DO"]]][["DO"]],
              key = list(version, as.character(db_species["DO"]), "DO"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["DO"],
                "-DO"
              )
            )
          }
        }

        ## HPO -----------------
        if (any(db == "HPO") && (!"HPO" %in% names(db_list[[sps]]))) {
          log_message("Preparing database: HPO")
          if (!sps %in% c("Homo_sapiens")) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the HPO database for ",
                sps,
                message_type = "warning"
              )
              db_species["HPO"] <- "Homo_sapiens"
            } else {
              log_message(
                "HPO database only support Homo_sapiens. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          temp <- tempfile()
          download(
            url = "https://api.github.com/repos/obophenotype/human-phenotype-ontology/releases?per_page=1",
            destfile = temp
          )
          release <- readLines(temp, warn = FALSE)
          version <- regmatches(
            release,
            m = regexpr(
              "(?<=tag_name\\\":\\\")\\S+(?=\\\",\\\"target_commitish)",
              release,
              perl = T
            )
          )

          download(
            url = "http://purl.obolibrary.org/obo/hp/hpoa/phenotype_to_genes.txt",
            destfile = temp
          )
          hpo <- utils::read.table(
            temp,
            header = TRUE,
            sep = "\t",
            fill = TRUE,
            quote = ""
          )
          unlink(temp)

          TERM2GENE <- hpo[, c("hpo_id", "gene_symbol")]
          TERM2NAME <- hpo[, c("hpo_id", "hpo_name")]
          colnames(TERM2GENE) <- c("Term", default_id_types[["HPO"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["HPO"]]][["HPO"]][["TERM2GENE"]] <- TERM2GENE
          db_list[[db_species["HPO"]]][["HPO"]][["TERM2NAME"]] <- TERM2NAME
          db_list[[db_species["HPO"]]][["HPO"]][["version"]] <- version
          if (sps == db_species["HPO"]) {
            R.cache::saveCache(
              db_list[[db_species["HPO"]]][["HPO"]],
              key = list(version, as.character(db_species["HPO"]), "HPO"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["HPO"],
                "-HPO"
              )
            )
          }
        }

        ## PFAM -----------------
        if (any(db == "PFAM") && (!"PFAM" %in% names(db_list[[sps]]))) {
          log_message("Preparing database: PFAM")
          if (!"PFAM" %in% AnnotationDbi::columns(orgdb)) {
            log_message(
              "PFAM is not in the orgdb: ",
              orgdb,
              " . Skip the preparation.",
              message_type = "warning"
            )
          } else {
            bg <- suppressMessages(
              AnnotationDbi::select(
                orgdb,
                keys = AnnotationDbi::keys(orgdb),
                columns = c("PFAM", org_key)
              )
            )
            bg <- unique(bg[!is.na(bg$PFAM), c("PFAM", org_key), drop = FALSE])
            bg2 <- as.data.frame(
              PFAM.db::PFAMDE2AC[AnnotationDbi::mappedkeys(PFAM.db::PFAMDE2AC)]
            )
            rownames(bg2) <- bg2[["ac"]]
            bg[["PFAM_name"]] <- bg2[bg$PFAM, "de"]
            bg[is.na(bg[["PFAM_name"]]), "PFAM_name"] <- bg[
              is.na(bg[["PFAM_name"]]),
              "PFAM"
            ]
            TERM2GENE <- bg[, c("PFAM", org_key)]
            TERM2NAME <- bg[, c("PFAM", "PFAM_name")]
            colnames(TERM2GENE) <- c("Term", default_id_types[["PFAM"]])
            colnames(TERM2NAME) <- c("Term", "Name")
            TERM2GENE <- stats::na.omit(unique(TERM2GENE))
            TERM2NAME <- stats::na.omit(unique(TERM2NAME))
            version <- utils::packageVersion(org_sp)
            db_list[[db_species["PFAM"]]][["PFAM"]][["TERM2GENE"]] <- TERM2GENE
            db_list[[db_species["PFAM"]]][["PFAM"]][["TERM2NAME"]] <- TERM2NAME
            db_list[[db_species["PFAM"]]][["PFAM"]][["version"]] <- version
            if (sps == db_species["PFAM"]) {
              R.cache::saveCache(
                db_list[[db_species["PFAM"]]][["PFAM"]],
                key = list(version, as.character(db_species["PFAM"]), "PFAM"),
                comment = paste0(
                  version,
                  " nterm:",
                  length(TERM2NAME[[1]]),
                  "|",
                  db_species["PFAM"],
                  "-PFAM"
                )
              )
            }
          }
        }

        ## Chromosome -----------------
        if (
          any(db == "Chromosome") && (!"Chromosome" %in% names(db_list[[sps]]))
        ) {
          log_message("Preparing database: Chromosome")
          orgdbCHR <- get(
            paste0(gsub(pattern = ".db", "", org_sp), "CHR")
          )
          chr <- as.data.frame(
            orgdbCHR[AnnotationDbi::mappedkeys(orgdbCHR)]
          )
          chr[, 2] <- paste0("chr", chr[, 2])
          TERM2GENE <- chr[, c(2, 1)]
          TERM2NAME <- chr[, c(2, 2)]
          colnames(TERM2GENE) <- c("Term", default_id_types[["Chromosome"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          version <- utils::packageVersion(org_sp)
          db_list[[db_species["Chromosome"]]][["Chromosome"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["Chromosome"]]][["Chromosome"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["Chromosome"]]][["Chromosome"]][[
            "version"
          ]] <- version
          if (sps == db_species["Chromosome"]) {
            R.cache::saveCache(
              db_list[[db_species["Chromosome"]]][["Chromosome"]],
              key = list(
                version,
                as.character(db_species["Chromosome"]),
                "Chromosome"
              ),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["Chromosome"],
                "-Chromosome"
              )
            )
          }
        }

        ## GeneType -----------------
        if (any(db == "GeneType") && (!"GeneType" %in% names(db_list[[sps]]))) {
          log_message("Preparing database: GeneType")
          if (!"GENETYPE" %in% AnnotationDbi::columns(orgdb)) {
            log_message(
              "GENETYPE is not in the orgdb: ",
              org_sp,
              " . Skip the preparation.",
              message_type = "warning"
            )
          } else {
            bg <- suppressMessages(
              AnnotationDbi::select(
                orgdb,
                keys = AnnotationDbi::keys(orgdb),
                columns = c("GENETYPE", org_key)
              )
            )
            TERM2GENE <- bg[, c("GENETYPE", org_key)]
            TERM2NAME <- bg[, c("GENETYPE", "GENETYPE")]
            colnames(TERM2GENE) <- c("Term", default_id_types[["GeneType"]])
            colnames(TERM2NAME) <- c("Term", "Name")
            TERM2GENE <- stats::na.omit(unique(TERM2GENE))
            TERM2NAME <- stats::na.omit(unique(TERM2NAME))
            version <- utils::packageVersion(org_sp)
            db_list[[db_species["GeneType"]]][["GeneType"]][[
              "TERM2GENE"
            ]] <- TERM2GENE
            db_list[[db_species["GeneType"]]][["GeneType"]][[
              "TERM2NAME"
            ]] <- TERM2NAME
            db_list[[db_species["GeneType"]]][["GeneType"]][[
              "version"
            ]] <- version
            if (sps == db_species["GeneType"]) {
              R.cache::saveCache(
                db_list[[db_species["GeneType"]]][["GeneType"]],
                key = list(
                  version,
                  as.character(db_species["GeneType"]),
                  "GeneType"
                ),
                comment = paste0(
                  version,
                  " nterm:",
                  length(TERM2NAME[[1]]),
                  "|",
                  db_species["GeneType"],
                  "-GeneType"
                )
              )
            }
          }
        }

        ## Enzyme -----------------
        if (any(db == "Enzyme") && (!"Enzyme" %in% names(db_list[[sps]]))) {
          log_message("Preparing database: Enzyme")
          if (!"ENZYME" %in% AnnotationDbi::columns(orgdb)) {
            log_message(
              "ENZYME is not in the orgdb: ",
              orgdb,
              " . Skip the preparation.",
              message_type = "warning"
            )
          } else {
            bg <- suppressMessages(
              AnnotationDbi::select(
                orgdb,
                keys = AnnotationDbi::keys(orgdb),
                columns = c("ENZYME", org_key)
              )
            )
            bg1 <- bg2 <- stats::na.omit(bg)
            bg1[, "ENZYME"] <- sapply(
              strsplit(bg1[, "ENZYME"], "\\."),
              function(x) paste0(utils::head(x, 1), collapse = ".")
            )
            bg2[, "ENZYME"] <- sapply(
              strsplit(bg2[, "ENZYME"], "\\."),
              function(x) paste0(utils::head(x, 2), collapse = ".")
            )
            bg <- unique(rbind(bg1, bg2))
            bg[, "ENZYME"] <- gsub(pattern = "\\.-$", "", x = bg[, 2])
            bg[, "ENZYME"] <- paste0("ec:", bg[, "ENZYME"])
            temp <- tempfile()
            download(
              url = "https://ftp.expasy.org/databases/enzyme/enzclass.txt",
              destfile = temp
            )
            enzyme <- utils::read.table(
              temp,
              header = FALSE,
              sep = "\t",
              fill = TRUE,
              quote = ""
            )
            enzyme <- enzyme[
              grep("-.-", enzyme[, 1], fixed = TRUE), ,
              drop = FALSE
            ]
            enzyme <- do.call(rbind, strsplit(enzyme[, 1], split = ". -.-  "))
            enzyme[, 1] <- paste0(
              "ec:",
              gsub(pattern = "( )|(. -)", replacement = "", enzyme[, 1])
            )
            enzyme[, 2] <- gsub(
              pattern = "(^ )|(\\.$)",
              replacement = "",
              enzyme[, 2]
            )
            rownames(enzyme) <- enzyme[, 1]
            for (i in seq_len(nrow(enzyme))) {
              if (grepl(".", enzyme[i, 1], fixed = TRUE)) {
                enzyme[i, 2] <- paste0(
                  enzyme[strsplit(enzyme[i, 1], ".", fixed = TRUE)[[1]][1], 2],
                  "(",
                  enzyme[i, 2],
                  ")"
                )
              }
            }
            unlink(temp)
            bg[, "Name"] <- enzyme[bg[, "ENZYME"], 2]
            TERM2GENE <- bg[, c("ENZYME", org_key)]
            TERM2NAME <- bg[, c("ENZYME", "Name")]
            colnames(TERM2GENE) <- c("Term", default_id_types[["Enzyme"]])
            colnames(TERM2NAME) <- c("Term", "Name")
            TERM2GENE <- stats::na.omit(unique(TERM2GENE))
            TERM2NAME <- stats::na.omit(unique(TERM2NAME))
            version <- utils::packageVersion(org_sp)
            db_list[[db_species["Enzyme"]]][["Enzyme"]][[
              "TERM2GENE"
            ]] <- TERM2GENE
            db_list[[db_species["Enzyme"]]][["Enzyme"]][[
              "TERM2NAME"
            ]] <- TERM2NAME
            db_list[[db_species["Enzyme"]]][["Enzyme"]][["version"]] <- version
            if (sps == db_species["Enzyme"]) {
              R.cache::saveCache(
                db_list[[db_species["Enzyme"]]][["Enzyme"]],
                key = list(
                  version,
                  as.character(db_species["Enzyme"]),
                  "Enzyme"
                ),
                comment = paste0(
                  version,
                  " nterm:",
                  length(TERM2NAME[[1]]),
                  "|",
                  db_species["Enzyme"],
                  "-Enzyme"
                )
              )
            }
          }
        }

        ## TF -----------------
        if (any(db == "TF") && (!"TF" %in% names(db_list[[sps]]))) {
          log_message("Preparing database: TF")

          # AnimalTFDB4
          status <- tryCatch(
            {
              temp <- tempfile()
              url <- paste0(
                "https://guolab.wchscu.cn/AnimalTFDB4_static/download/TF_list_final/",
                sps,
                "_TF"
              )
              download(url = url, destfile = temp, use_httr = TRUE)
              tf <- utils::read.table(
                temp,
                header = TRUE,
                sep = "\t",
                stringsAsFactors = FALSE,
                fill = TRUE,
                quote = ""
              )
              url <- paste0(
                "https://guolab.wchscu.cn/AnimalTFDB4_static/download/Cof_list_final/",
                sps,
                "_Cof"
              )
              download(url = url, destfile = temp, use_httr = TRUE)
              tfco <- utils::read.table(
                temp,
                header = TRUE,
                sep = "\t",
                stringsAsFactors = FALSE,
                fill = TRUE,
                quote = ""
              )
              if (!"Symbol" %in% colnames(tf)) {
                if (
                  isTRUE(convert_species) && db_species["TF"] != "Homo_sapiens"
                ) {
                  log_message(
                    "Use the human annotation to create the TF database for ",
                    sps,
                    message_type = "warning"
                  )
                  db_species["TF"] <- "Homo_sapiens"
                  url <- paste0(
                    "https://guolab.wchscu.cn/AnimalTFDB4_static/download/TF_list_final/Homo_sapiens_TF"
                  )
                  download(url = url, destfile = temp, use_httr = TRUE)
                  tf <- utils::read.table(
                    temp,
                    header = TRUE,
                    sep = "\t",
                    stringsAsFactors = FALSE,
                    fill = TRUE,
                    quote = ""
                  )
                  url <- paste0(
                    "https://guolab.wchscu.cn/AnimalTFDB4_static/download/Cof_list_final/Homo_sapiens_Cof"
                  )
                  download(url = url, destfile = temp, use_httr = TRUE)
                  tfco <- utils::read.table(
                    temp,
                    header = TRUE,
                    sep = "\t",
                    stringsAsFactors = FALSE,
                    fill = TRUE,
                    quote = ""
                  )
                } else {
                  log_message(
                    "Stop the preparation.",
                    message_type = "error"
                  )
                }
              }
              unlink(temp)
              version <- "AnimalTFDB4"
            },
            error = identity
          )

          # AnimalTFDB3
          if (inherits(status, "error")) {
            temp <- tempfile()
            url <- paste0(
              "https://raw.githubusercontent.com/GuoBioinfoLab/AnimalTFDB3/master/AnimalTFDB3/static/AnimalTFDB3/download/",
              sps,
              "_TF"
            )
            download(url = url, destfile = temp)
            tf <- utils::read.table(
              temp,
              header = TRUE,
              sep = "\t",
              stringsAsFactors = FALSE,
              fill = TRUE,
              quote = ""
            )
            url <- paste0(
              "https://raw.githubusercontent.com/GuoBioinfoLab/AnimalTFDB3/master/AnimalTFDB3/static/AnimalTFDB3/download/",
              sps,
              "_TF_cofactors"
            )
            download(url = url, destfile = temp)
            tfco <- utils::read.table(
              temp,
              header = TRUE,
              sep = "\t",
              stringsAsFactors = FALSE,
              fill = TRUE,
              quote = ""
            )
            if (!"Symbol" %in% colnames(tf)) {
              if (isTRUE(convert_species) && db_species["TF"] != "Homo_sapiens") {
                log_message(
                  "Use the human annotation to create the TF database for ",
                  sps,
                  message_type = "warning"
                )
                db_species["TF"] <- "Homo_sapiens"
                url <- c(
                  "https://raw.githubusercontent.com/GuoBioinfoLab/AnimalTFDB3/master/AnimalTFDB3/static/AnimalTFDB3/download/Homo_sapiens_TF"
                )
                download(url = url, destfile = temp)
                tf <- utils::read.table(
                  temp,
                  header = TRUE,
                  sep = "\t",
                  stringsAsFactors = FALSE,
                  fill = TRUE,
                  quote = ""
                )
                url <- paste0(
                  "https://raw.githubusercontent.com/GuoBioinfoLab/AnimalTFDB3/master/AnimalTFDB3/static/AnimalTFDB3/download/Homo_sapiens_TF_cofactors"
                )
                download(url = url, destfile = temp)
                tfco <- utils::read.table(
                  temp,
                  header = TRUE,
                  sep = "\t",
                  stringsAsFactors = FALSE,
                  fill = TRUE,
                  quote = ""
                )
              } else {
                log_message(
                  "Stop the preparation.",
                  message_type = "error"
                )
              }
            }
            unlink(temp)
            version <- "AnimalTFDB3"
          }

          TERM2GENE <- rbind(
            data.frame("Term" = "TF", "symbol" = tf[["Symbol"]]),
            data.frame("Term" = "TF cofactor", "symbol" = tfco[["Symbol"]])
          )
          TERM2NAME <- data.frame(
            "Term" = c("TF", "TF cofactor"),
            "Name" = c("TF", "TF cofactor")
          )
          colnames(TERM2GENE) <- c("Term", default_id_types[["TF"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["TF"]]][["TF"]][["TERM2GENE"]] <- TERM2GENE
          db_list[[db_species["TF"]]][["TF"]][["TERM2NAME"]] <- TERM2NAME
          db_list[[db_species["TF"]]][["TF"]][["version"]] <- version
          if (sps == db_species["TF"]) {
            R.cache::saveCache(
              db_list[[db_species["TF"]]][["TF"]],
              key = list(version, as.character(db_species["TF"]), "TF"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["TF"],
                "-TF"
              )
            )
          }
        }

        ## CSPA -----------------
        if (any(db == "CSPA") && (!"CSPA" %in% names(db_list[[sps]]))) {
          if (!sps %in% c("Homo_sapiens", "Mus_musculus")) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the CSPA database for ",
                sps,
                message_type = "warning"
              )
              db_species["CSPA"] <- "Homo_sapiens"
            } else {
              log_message(
                "CSPA database only support Homo_sapiens and Mus_musculus. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          check_r("openxlsx")
          log_message("Preparing database: CSPA")
          temp <- tempfile(fileext = ".xlsx")
          url <- "https://wlab.ethz.ch/cspa/data/S1_File.xlsx"
          download(
            url = url,
            destfile = temp,
            mode = ifelse(.Platform$OS.type == "windows", "wb", "w")
          )
          surfacepro <- openxlsx::read.xlsx(temp, sheet = 1)
          unlink(temp)
          surfacepro <- surfacepro[
            surfacepro[["organism"]] ==
              switch(db_species["CSPA"],
                "Homo_sapiens" = "Human",
                "Mus_musculus" = "Mouse"
              ), ,
            drop = FALSE
          ]
          TERM2GENE <- data.frame(
            "Term" = "SurfaceProtein",
            "symbol" = surfacepro[["ENTREZ.gene.symbol"]]
          )
          TERM2NAME <- data.frame(
            "Term" = "SurfaceProtein",
            "Name" = "SurfaceProtein"
          )
          colnames(TERM2GENE) <- c("Term", default_id_types[["CSPA"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          version <- "CSPA"
          db_list[[db_species["CSPA"]]][["CSPA"]][["TERM2GENE"]] <- TERM2GENE
          db_list[[db_species["CSPA"]]][["CSPA"]][["TERM2NAME"]] <- TERM2NAME
          db_list[[db_species["CSPA"]]][["CSPA"]][["version"]] <- version
          if (sps == db_species["CSPA"]) {
            R.cache::saveCache(
              db_list[[db_species["CSPA"]]][["CSPA"]],
              key = list(version, as.character(db_species["CSPA"]), "CSPA"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["CSPA"],
                "-CSPA"
              )
            )
          }
        }

        ## Surfaceome -----------------
        if (
          any(db == "Surfaceome") && (!"Surfaceome" %in% names(db_list[[sps]]))
        ) {
          if (!sps %in% c("Homo_sapiens")) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the Surfaceome database for ",
                sps,
                message_type = "warning"
              )
              db_species["Surfaceome"] <- "Homo_sapiens"
            } else {
              log_message(
                "Surfaceome database only support Homo_sapiens. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          check_r("openxlsx")
          log_message("Preparing database: Surfaceome")
          temp <- tempfile(fileext = ".xlsx")
          url <- "http://wlab.ethz.ch/surfaceome/table_S3_surfaceome.xlsx"
          download(
            url = url,
            destfile = temp,
            mode = ifelse(.Platform$OS.type == "windows", "wb", "w")
          )
          surfaceome <- openxlsx::read.xlsx(
            temp,
            sheet = 2,
            colNames = TRUE,
            startRow = 2
          )
          unlink(temp)
          TERM2GENE <- data.frame(
            "Term" = "SurfaceProtein",
            "symbol" = surfaceome[["UniProt.gene"]]
          )
          TERM2NAME <- data.frame(
            "Term" = "SurfaceProtein",
            "Name" = "SurfaceProtein"
          )
          colnames(TERM2GENE) <- c("Term", default_id_types[["Surfaceome"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          version <- "Surfaceome"
          db_list[[db_species["Surfaceome"]]][["Surfaceome"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["Surfaceome"]]][["Surfaceome"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["Surfaceome"]]][["Surfaceome"]][[
            "version"
          ]] <- version
          if (sps == db_species["Surfaceome"]) {
            R.cache::saveCache(
              db_list[[db_species["Surfaceome"]]][["Surfaceome"]],
              key = list(
                version,
                as.character(db_species["Surfaceome"]),
                "Surfaceome"
              ),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["Surfaceome"],
                "-Surfaceome"
              )
            )
          }
        }

        ## SPRomeDB -----------------
        if (any(db == "SPRomeDB") && (!"SPRomeDB" %in% names(db_list[[sps]]))) {
          if (!sps %in% c("Homo_sapiens")) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the SPRomeDB database for ",
                sps,
                message_type = "warning"
              )
              db_species["SPRomeDB"] <- "Homo_sapiens"
            } else {
              log_message(
                "SPRomeDB database only support Homo_sapiens. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: SPRomeDB")
          temp <- tempfile()
          url <- "http://119.3.41.228/SPRomeDB/files/download/secreted_proteins_SPRomeDB.csv"
          download(url = url, destfile = temp)
          spromedb <- utils::read.csv(temp, header = TRUE)
          unlink(temp)
          TERM2GENE <- data.frame(
            "Term" = "SecretoryProtein",
            "entrez_id" = unlist(strsplit(spromedb$Gene_ID, ";"))
          )
          TERM2NAME <- data.frame(
            "Term" = "SecretoryProtein",
            "Name" = "SecretoryProtein"
          )
          colnames(TERM2GENE) <- c("Term", default_id_types[["SPRomeDB"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          version <- "SPRomeDB"
          db_list[[db_species["SPRomeDB"]]][["SPRomeDB"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["SPRomeDB"]]][["SPRomeDB"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["SPRomeDB"]]][["SPRomeDB"]][[
            "version"
          ]] <- version
          if (sps == db_species["SPRomeDB"]) {
            R.cache::saveCache(
              db_list[[db_species["SPRomeDB"]]][["SPRomeDB"]],
              key = list(
                version,
                as.character(db_species["SPRomeDB"]),
                "SPRomeDB"
              ),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["SPRomeDB"],
                "-SPRomeDB"
              )
            )
          }
        }

        ## VerSeDa -----------------
        if (any(db == "VerSeDa") && (!"VerSeDa" %in% names(db_list[[sps]]))) {
          temp <- tempfile()
          download(
            url = "http://genomics.cicbiogune.es/VerSeDa/downloads.php",
            destfile = temp
          )
          verseda_sps <- readLines(temp)
          verseda_sps <- regmatches(
            verseda_sps,
            m = regexpr(
              "(?<=Downloads/)\\S+(?=\\.zip)",
              verseda_sps,
              perl = TRUE
            )
          )
          verseda_sps <- setdiff(
            verseda_sps,
            c("NonRefined", "NonRefined_Curated", "Refined", "Refined_Curated")
          )
          if (!tolower(sps) %in% verseda_sps) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the VerSeDa database for ",
                sps,
                message_type = "warning"
              )
              db_species["VerSeDa"] <- "Homo_sapiens"
            } else {
              log_message(
                "VerSeDa database only support ",
                paste0(verseda_sps, collapse = ","),
                ". Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: VerSeDa")
          temp <- tempfile(fileext = ".zip")
          url <- paste0(
            "http://genomics.cicbiogune.es/VerSeDa/Downloads/",
            tolower(db_species["VerSeDa"]),
            ".zip"
          )
          download(url = url, destfile = temp)
          con <- unz(
            temp,
            paste0(
              tolower(db_species["VerSeDa"]),
              "/",
              tolower(db_species["VerSeDa"]),
              "_Refined.sequences"
            )
          )
          verseda <- readLines(con)
          close(con)
          unlink(temp)
          verseda <- verseda[grep("^>", verseda)]
          verseda <- gsub("^>|\\.\\d+", "", verseda)
          verseda_id <- GeneConvert(
            geneID = verseda,
            geneID_from_IDtype = c(
              "ensembl_peptide_id",
              "refseq_peptide",
              "refseq_peptide_predicted",
              "uniprot_isoform",
              "uniprotswissprot",
              "uniprotsptrembl"
            ),
            geneID_to_IDtype = "symbol",
            species_from = db_species["VerSeDa"]
          )
          TERM2GENE <- data.frame(
            "Term" = "SecretoryProtein",
            "symbol" = unique(verseda_id$geneID_expand$symbol)
          )
          TERM2NAME <- data.frame(
            "Term" = "SecretoryProtein",
            "Name" = "SecretoryProtein"
          )
          colnames(TERM2GENE) <- c("Term", default_id_types[["VerSeDa"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          version <- "VerSeDa"
          db_list[[db_species["VerSeDa"]]][["VerSeDa"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["VerSeDa"]]][["VerSeDa"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["VerSeDa"]]][["VerSeDa"]][["version"]] <- version
          if (sps == db_species["VerSeDa"]) {
            R.cache::saveCache(
              db_list[[db_species["VerSeDa"]]][["VerSeDa"]],
              key = list(
                version,
                as.character(db_species["VerSeDa"]),
                "VerSeDa"
              ),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["VerSeDa"],
                "-VerSeDa"
              )
            )
          }
        }

        ## TFLink -----------------
        if (any(db == "TFLink") && (!"TFLink" %in% names(db_list[[sps]]))) {
          tflink_sp <- c(
            "Homo_sapiens",
            "Mus_musculus",
            "Rattus_norvegicus",
            "Danio_rerio",
            "Drosophila_melanogaster",
            "Caenorhabditis_elegans",
            "Saccharomyces_cerevisiae"
          )
          if (!sps %in% tflink_sp) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the TFLink database for ",
                sps,
                message_type = "warning"
              )
              db_species["TFLink"] <- "Homo_sapiens"
            } else {
              log_message(
                "TFLink database only support ",
                paste0(tflink_sp, collapse = ","),
                ". Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: TFLink")
          url <- paste0(
            "https://cdn.netbiol.org/tflink/download_files/TFLink_",
            db_species["TFLink"],
            "_interactions_All_GMT_proteinName_v1.0.gmt"
          )
          temp <- tempfile()
          download(url = url, destfile = temp)
          TERM2GENE <- clusterProfiler::read.gmt(temp)
          version <- "v1.0"
          TERM2NAME <- TERM2GENE[, c(1, 1)]
          colnames(TERM2GENE) <- c("Term", default_id_types[["TFLink"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["TFLink"]]][["TFLink"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["TFLink"]]][["TFLink"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["TFLink"]]][["TFLink"]][["version"]] <- version
          if (sps == db_species["TFLink"]) {
            R.cache::saveCache(
              db_list[[db_species["TFLink"]]][["TFLink"]],
              key = list(version, as.character(db_species["TFLink"]), "TFLink"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["TFLink"],
                "-TFLink"
              )
            )
          }
        }

        ## hTFtarget -----------------
        if (
          any(db == "hTFtarget") && (!"hTFtarget" %in% names(db_list[[sps]]))
        ) {
          if (!sps %in% "Homo_sapiens") {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the hTFtarget database for ",
                sps,
                message_type = "warning"
              )
              db_species["hTFtarget"] <- "Homo_sapiens"
            } else {
              log_message(
                "hTFtarget database only support Homo_sapiens. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: hTFtarget")
          url <- paste0(
            "http://bioinfo.life.hust.edu.cn/static/hTFtarget/file_download/tf-target-infomation.txt"
          )
          temp <- tempfile()
          download(url = url, destfile = temp)
          TERM2GENE <- utils::read.table(temp, header = TRUE, fill = T, sep = "\t")
          version <- "v1.0"
          TERM2NAME <- TERM2GENE[, c(1, 1)]
          colnames(TERM2GENE) <- c("Term", default_id_types[["hTFtarget"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["hTFtarget"]]][["hTFtarget"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["hTFtarget"]]][["hTFtarget"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["hTFtarget"]]][["hTFtarget"]][[
            "version"
          ]] <- version
          if (sps == db_species["hTFtarget"]) {
            R.cache::saveCache(
              db_list[[db_species["hTFtarget"]]][["hTFtarget"]],
              key = list(
                version,
                as.character(db_species["hTFtarget"]),
                "hTFtarget"
              ),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["hTFtarget"],
                "-hTFtarget"
              )
            )
          }
        }

        ## TRRUST -----------------
        if (any(db == "TRRUST") && (!"TRRUST" %in% names(db_list[[sps]]))) {
          if (!sps %in% c("Homo_sapiens", "Mus_musculus")) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the TRRUST database for ",
                sps,
                message_type = "warning"
              )
              db_species["TRRUST"] <- "Homo_sapiens"
            } else {
              log_message(
                "hTFtarget database only support Homo_sapiens and Mus_musculus. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: TRRUST")
          url <- switch(db_species["TRRUST"],
            "Homo_sapiens" = "https://raw.githubusercontent.com/bioinfonerd/Transcription-Factor-Databases/master/Ttrust_v2/trrust_rawdata.human.tsv",
            "Mus_musculus" = "https://raw.githubusercontent.com/bioinfonerd/Transcription-Factor-Databases/master/Ttrust_v2/trrust_rawdata.mouse.tsv.gz"
          )
          if (endsWith(url, "gz")) {
            temp <- tempfile(fileext = ".gz")
            download(url = url, destfile = temp)
            R.utils::gunzip(temp)
            TERM2GENE <- utils::read.table(
              gsub(".gz$", "", temp),
              header = FALSE,
              fill = T,
              sep = "\t"
            )[, 1:2]
          } else {
            temp <- tempfile()
            download(url = url, destfile = temp)
            TERM2GENE <- utils::read.table(
              temp,
              header = FALSE,
              fill = T,
              sep = "\t"
            )[, 1:2]
          }
          version <- "v2.0"
          TERM2NAME <- TERM2GENE[, c(1, 1)]
          colnames(TERM2GENE) <- c("Term", default_id_types[["TRRUST"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["TRRUST"]]][["TRRUST"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["TRRUST"]]][["TRRUST"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["TRRUST"]]][["TRRUST"]][["version"]] <- version
          if (sps == db_species["TRRUST"]) {
            R.cache::saveCache(
              db_list[[db_species["TRRUST"]]][["TRRUST"]],
              key = list(version, as.character(db_species["TRRUST"]), "TRRUST"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["TRRUST"],
                "-TRRUST"
              )
            )
          }
        }

        ## JASPAR -----------------
        if (any(db == "JASPAR") && (!"JASPAR" %in% names(db_list[[sps]]))) {
          if (!sps %in% c("Homo_sapiens")) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the JASPAR database for ",
                sps,
                message_type = "warning"
              )
              db_species["JASPAR"] <- "Homo_sapiens"
            } else {
              log_message(
                "JASPAR database only support Homo_sapiens. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: JASPAR")
          url <- "https://maayanlab.cloud/static/hdfs/harmonizome/data/jasparpwm/gene_set_library_crisp.gmt.gz"
          temp <- tempfile(fileext = ".gz")
          download(url = url, destfile = temp)
          R.utils::gunzip(temp)
          TERM2GENE <- clusterProfiler::read.gmt(gsub(".gz", "", temp))
          version <- "Harmonizome 3.0"
          TERM2NAME <- TERM2GENE[, c(1, 1)]
          colnames(TERM2GENE) <- c("Term", default_id_types[["JASPAR"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["JASPAR"]]][["JASPAR"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["JASPAR"]]][["JASPAR"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["JASPAR"]]][["JASPAR"]][["version"]] <- version
          if (sps == db_species["JASPAR"]) {
            R.cache::saveCache(
              db_list[[db_species["JASPAR"]]][["JASPAR"]],
              key = list(version, as.character(db_species["JASPAR"]), "JASPAR"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["JASPAR"],
                "-JASPAR"
              )
            )
          }
        }

        ## ENCODE -----------------
        if (any(db == "ENCODE") && (!"ENCODE" %in% names(db_list[[sps]]))) {
          if (!sps %in% c("Homo_sapiens")) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the ENCODE database for ",
                sps,
                message_type = "warning"
              )
              db_species["ENCODE"] <- "Homo_sapiens"
            } else {
              log_message(
                "ENCODE database only support Homo_sapiens. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: ENCODE")
          url <- "https://maayanlab.cloud/static/hdfs/harmonizome/data/encodetfppi/gene_set_library_crisp.gmt.gz"
          temp <- tempfile(fileext = ".gz")
          download(url = url, destfile = temp)
          R.utils::gunzip(temp)
          TERM2GENE <- clusterProfiler::read.gmt(gsub(".gz", "", temp))
          version <- "Harmonizome 3.0"
          TERM2NAME <- TERM2GENE[, c(1, 1)]
          colnames(TERM2GENE) <- c("Term", default_id_types[["ENCODE"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["ENCODE"]]][["ENCODE"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["ENCODE"]]][["ENCODE"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["ENCODE"]]][["ENCODE"]][["version"]] <- version
          if (sps == db_species["ENCODE"]) {
            R.cache::saveCache(
              db_list[[db_species["ENCODE"]]][["ENCODE"]],
              key = list(version, as.character(db_species["ENCODE"]), "ENCODE"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["ENCODE"],
                "-ENCODE"
              )
            )
          }
        }

        ## MSigDB -----------------
        if (
          any(grepl("MSigDB", db)) && (!"MSigDB" %in% names(db_list[[sps]]))
        ) {
          if (!sps %in% c("Homo_sapiens", "Mus_musculus")) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the MSigDB database for ",
                sps,
                message_type = "warning"
              )
              db_species["MSigDB"] <- "Homo_sapiens"
            } else {
              log_message(
                "MSigDB database only support Homo_sapiens and Mus_musculus. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: MSigDB")

          temp <- tempfile()
          download(
            url = "https://data.broadinstitute.org/gsea-msigdb/msigdb/release/",
            destfile = temp
          )
          version <- readLines(temp)
          version <- version[grep("alt=\"\\[DIR\\]\"", version)]
          version <- version[grep(
            switch(db_species["MSigDB"],
              "Homo_sapiens" = "Hs",
              "Mus_musculus" = "Mm"
            ),
            version
          )]
          version <- version[length(version)]
          version <- regmatches(
            version,
            m = regexpr("(?<=href\\=\")\\S+(?=/\"\\>)", version, perl = TRUE)
          )
          # version="2023.1.Hs" "2023.2.Hs"

          url <- paste0(
            "https://data.broadinstitute.org/gsea-msigdb/msigdb/release/",
            version,
            "/msigdb.v",
            version,
            ".json"
          )
          download(url = url, destfile = temp)
          lines <- paste0(readLines(temp, warn = FALSE), collapse = "")
          lines <- gsub("\"", "", lines)
          lines <- gsub("^\\{|\\}$", "", lines)
          terms <- strsplit(lines, "\\},")[[1]]
          term_list <- list()
          for (idx in seq_along(terms)) {
            term_content <- terms[[idx]]
            term_name <- trimws(gsub(":|_|\\{.*", " ", term_content))
            term_id <- trimws(regmatches(
              term_content,
              m = regexpr(
                "(?<=systematicName:)\\S+?(?=,)",
                term_content,
                perl = TRUE
              )
            ))
            term_gene <- trimws(regmatches(
              term_content,
              m = regexpr(
                pattern = "(?<=geneSymbols:\\[)\\S+?(?=\\],)",
                term_content,
                perl = TRUE
              )
            ))
            term_collection <- trimws(regmatches(
              term_content,
              m = regexpr(
                pattern = "(?<=collection:)\\S+?(?=\\,)",
                term_content,
                perl = TRUE
              )
            ))
            term_collection <- gsub(":.*", "", term_collection)
            term_list[[idx]] <- c(
              id = term_id,
              "name" = term_name,
              "gene" = term_gene,
              "collection" = term_collection
            )
          }
          df <- as.data.frame(do.call(rbind, term_list))
          df$gene <- strsplit(df$gene, split = ",")
          df <- unnest_fun(df, cols = "gene")

          TERM2NAME <- df[, c(1, 2, 4)]
          TERM2GENE <- df[, c(1, 3)]
          colnames(TERM2NAME) <- c("Term", "Name", "Collection")
          colnames(TERM2GENE) <- c(
            "Term",
            paste0(default_id_types[["MSigDB"]], collapse = ".")
          )
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))

          db_list[[db_species["MSigDB"]]][["MSigDB"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["MSigDB"]]][["MSigDB"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["MSigDB"]]][["MSigDB"]][["version"]] <- version
          if (sps == db_species["MSigDB"]) {
            R.cache::saveCache(
              db_list[[db_species["MSigDB"]]][["MSigDB"]],
              key = list(version, as.character(db_species["MSigDB"]), "MSigDB"),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["MSigDB"],
                "-MSigDB"
              )
            )
          }

          for (collection in unique(TERM2NAME[["Collection"]])) {
            db_species[paste0("MSigDB_", collection)] <- db_species["MSigDB"]
            default_id_types[[paste0(
              "MSigDB_",
              collection
            )]] <- default_id_types[["MSigDB"]]
            TERM2NAME_sub <- TERM2NAME[
              TERM2NAME[["Collection"]] == collection, ,
              drop = FALSE
            ]
            TERM2GENE_sub <- TERM2GENE[
              TERM2GENE[["Term"]] %in% TERM2NAME_sub[["Term"]], ,
              drop = FALSE
            ]
            db_list[[db_species["MSigDB"]]][[paste0("MSigDB_", collection)]][[
              "TERM2GENE"
            ]] <- TERM2GENE_sub
            db_list[[db_species["MSigDB"]]][[paste0("MSigDB_", collection)]][[
              "TERM2NAME"
            ]] <- TERM2NAME_sub
            db_list[[db_species["MSigDB"]]][[paste0("MSigDB_", collection)]][[
              "version"
            ]] <- version
            if (sps == db_species["MSigDB"]) {
              R.cache::saveCache(
                db_list[[db_species["MSigDB"]]][[paste0(
                  "MSigDB_",
                  collection
                )]],
                key = list(
                  version,
                  as.character(db_species["MSigDB"]),
                  paste0("MSigDB_", collection)
                ),
                comment = paste0(
                  version,
                  " nterm:",
                  length(TERM2NAME[[1]]),
                  "|",
                  db_species["MSigDB"],
                  "-MSigDB_",
                  collection
                )
              )
            }
          }
        }

        ## CellTalk -----------------
        if (any(db == "CellTalk") && (!"CellTalk" %in% names(db_list[[sps]]))) {
          if (!sps %in% c("Homo_sapiens", "Mus_musculus")) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the CellTalk database for ",
                sps,
                message_type = "warning"
              )
              db_species["CellTalk"] <- "Homo_sapiens"
            } else {
              log_message(
                "CellTalk database only support Homo_sapiens and Mus_musculus. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: CellTalk")
          url <- switch(db_species["CellTalk"],
            "Homo_sapiens" = "https://raw.githubusercontent.com/ZJUFanLab/CellTalkDB/master/database/human_lr_pair.rds",
            "Mus_musculus" = "https://raw.githubusercontent.com/ZJUFanLab/CellTalkDB/master/database/mouse_lr_pair.rds"
          )

          temp <- tempfile()
          download(url = url, destfile = temp)
          lr <- readRDS(temp)
          version <- "v1.0"

          lr[["ligand_gene_symbol2"]] <- paste0(
            "ligand_",
            lr[["ligand_gene_symbol"]]
          )
          lr[["receptor_gene_symbol2"]] <- paste0(
            "receptor_",
            lr[["receptor_gene_symbol"]]
          )
          TERM2GENE <- rbind(
            data.frame(
              "Term" = lr[["ligand_gene_symbol2"]],
              "symbol" = lr[["receptor_gene_symbol"]]
            ),
            data.frame(
              "Term" = lr[["receptor_gene_symbol2"]],
              "symbol" = lr[["ligand_gene_symbol"]]
            )
          )
          TERM2NAME <- TERM2GENE[, c(1, 1)]
          colnames(TERM2GENE) <- c("Term", default_id_types[["CellTalk"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["CellTalk"]]][["CellTalk"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["CellTalk"]]][["CellTalk"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["CellTalk"]]][["CellTalk"]][[
            "version"
          ]] <- version
          if (sps == db_species["CellTalk"]) {
            R.cache::saveCache(
              db_list[[db_species["CellTalk"]]][["CellTalk"]],
              key = list(
                version,
                as.character(db_species["CellTalk"]),
                "CellTalk"
              ),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["CellTalk"],
                "-CellTalk"
              )
            )
          }
        }

        ## CellChat -----------------
        if (any(db == "CellChat") && (!"CellChat" %in% names(db_list[[sps]]))) {
          if (!sps %in% c("Homo_sapiens", "Mus_musculus")) {
            if (isTRUE(convert_species)) {
              log_message(
                "Use the human annotation to create the CellChat database for ",
                sps,
                message_type = "warning"
              )
              db_species["CellChat"] <- "Homo_sapiens"
            } else {
              log_message(
                "CellChat database only support Homo_sapiens and Mus_musculus. Consider using convert_species=TRUE",
                message_type = "warning"
              )
              log_message(
                "Stop the preparation.",
                message_type = "error"
              )
            }
          }
          log_message("Preparing database: CellChat")
          url <- paste0(
            "https://raw.githubusercontent.com/sqjin/CellChat/master/data/CellChatDB.",
            switch(db_species["CellChat"],
              "Homo_sapiens" = "human.rda",
              "Mus_musculus" = "mouse.rda",
              "Danio_rerio" = "zebrafish.rda"
            )
          )
          temp <- tempfile()
          download(url = url, destfile = temp)
          load(temp)
          lr <- get(paste0(
            "CellChatDB.",
            switch(db_species["CellChat"],
              "Homo_sapiens" = "human",
              "Mus_musculus" = "mouse",
              "Danio_rerio" = "zebrafish"
            )
          ))[["interaction"]]
          download(
            url = "https://raw.githubusercontent.com/sqjin/CellChat/master/DESCRIPTION",
            destfile = temp
          )
          version <- grep(
            pattern = "Version",
            x = readLines(temp),
            value = TRUE
          )
          version <- gsub(
            pattern = "(.*Version: )|(</td>)",
            replacement = "",
            x = version
          )
          unlink(temp)

          lr_list <- strsplit(lr$interaction_name, split = "_")
          lr[["ligand_gene_symbol"]] <- paste0(
            "ligand_",
            sapply(lr_list, function(x) x[[1]])
          )
          lr[["receptor_list"]] <- lapply(
            lr_list,
            function(x) paste0("receptor_", x[2:length(x)])
          )
          lr <- unnest_fun(data = lr, cols = "receptor_list", keep_empty = FALSE)
          TERM2GENE <- rbind(
            data.frame(
              "Term" = lr[["ligand_gene_symbol"]],
              "symbol" = gsub(
                pattern = "receptor_",
                replacement = "",
                lr[["receptor_list"]]
              )
            ),
            data.frame(
              "Term" = lr[["receptor_list"]],
              "symbol" = gsub(
                pattern = "ligand_",
                replacement = "",
                lr[["ligand_gene_symbol"]]
              )
            )
          )

          if (db_species["CellChat"] == "Homo_sapiens") {
            TERM2GENE[["symbol"]] <- toupper(TERM2GENE[["symbol"]])
          } else if (db_species["CellChat"] == "Mus_musculus") {
            TERM2GENE[["symbol"]] <- capitalize(
              TERM2GENE[["symbol"]],
              force_tolower = TRUE
            )
          } else if (db_species["CellChat"] == "Danio_rerio") {
            TERM2GENE[["symbol"]] <- tolower(TERM2GENE[["symbol"]])
          }
          TERM2NAME <- TERM2GENE[, c(1, 1)]
          colnames(TERM2GENE) <- c("Term", default_id_types[["CellChat"]])
          colnames(TERM2NAME) <- c("Term", "Name")
          TERM2GENE <- stats::na.omit(unique(TERM2GENE))
          TERM2NAME <- stats::na.omit(unique(TERM2NAME))
          db_list[[db_species["CellChat"]]][["CellChat"]][[
            "TERM2GENE"
          ]] <- TERM2GENE
          db_list[[db_species["CellChat"]]][["CellChat"]][[
            "TERM2NAME"
          ]] <- TERM2NAME
          db_list[[db_species["CellChat"]]][["CellChat"]][[
            "version"
          ]] <- version
          if (sps == db_species["CellChat"]) {
            R.cache::saveCache(
              db_list[[db_species["CellChat"]]][["CellChat"]],
              key = list(
                version,
                as.character(db_species["CellChat"]),
                "CellChat"
              ),
              comment = paste0(
                version,
                " nterm:",
                length(TERM2NAME[[1]]),
                "|",
                db_species["CellChat"],
                "-CellChat"
              )
            )
          }
        }
      } else {
        ## Custom -----------------
        db_species[db] <- custom_species
        if (sps != custom_species) {
          if (isTRUE(convert_species)) {
            log_message(
              "Use the ",
              custom_species,
              " annotation to create the ",
              db,
              " database for ",
              sps,
              message_type = "warning"
            )
          } else {
            log_message(
              db,
              " database only support ",
              custom_species,
              ". Consider using convert_species=TRUE",
              message_type = "warning"
            )
            log_message(
              "Stop the preparation.",
              message_type = "error"
            )
          }
        }
        TERM2GENE <- custom_TERM2GENE
        colnames(TERM2GENE) <- c("Term", custom_IDtype)
        if (is.null(custom_TERM2NAME)) {
          TERM2NAME <- TERM2GENE[, c(1, 1)]
        } else {
          TERM2NAME <- custom_TERM2NAME
        }
        colnames(TERM2NAME) <- c("Term", "Name")

        TERM2GENE <- stats::na.omit(unique(TERM2GENE))
        TERM2NAME <- stats::na.omit(unique(TERM2NAME))
        db_list[[db_species[db]]][[db]][["TERM2GENE"]] <- TERM2GENE
        db_list[[db_species[db]]][[db]][["TERM2NAME"]] <- TERM2NAME
        db_list[[db_species[db]]][[db]][["version"]] <- custom_version
        if (sps == db_species[db]) {
          R.cache::saveCache(
            db_list[[db_species[db]]][[db]],
            key = list(custom_version, as.character(db_species[db]), db),
            comment = paste0(
              custom_version,
              " nterm:",
              length(TERM2NAME[[1]]),
              "|",
              db_species[db],
              "-",
              db
            )
          )
        }
      }
    }

    # Convert species
    if (!all(db_species == sps)) {
      for (term in names(db_species[db_species != sps])) {
        log_message("Convert species for the database: ", term)
        sp_from <- db_species[term]
        db_info <- db_list[[sp_from]][[names(sp_from)]]
        TERM2GENE <- db_info[["TERM2GENE"]]
        TERM2NAME <- db_info[["TERM2NAME"]]
        if (is.na(default_id_types[term])) {
          IDtype <- colnames(TERM2GENE)[2]
        } else {
          IDtype <- default_id_types[[term]]
        }
        if (grepl("MSigDB_", term)) {
          TERM2GENE_map <- db_list[[sps]][["MSigDB"]][["TERM2GENE"]]
          TERM2GENE <- TERM2GENE_map[
            TERM2GENE_map[["Term"]] %in% TERM2GENE[["Term"]], ,
            drop = FALSE
          ]
          TERM2NAME <- TERM2NAME[
            TERM2NAME[["Term"]] %in% TERM2GENE[["Term"]], ,
            drop = FALSE
          ]
        } else {
          res <- GeneConvert(
            geneID = as.character(unique(TERM2GENE[, 2])),
            geneID_from_IDtype = IDtype,
            geneID_to_IDtype = "ensembl_id",
            species_from = sp_from,
            species_to = sps,
            Ensembl_version = Ensembl_version,
            mirror = mirror,
            biomart = biomart,
            max_tries = max_tries
          )
          if (is.null(res$geneID_res)) {
            log_message(
              "Failed to convert species for the database: ",
              term,
              message_type = "warning"
            )
            next
          }
          map <- res$geneID_collapse
          TERM2GENE[["ensembl_id-converted"]] <- map[
            as.character(TERM2GENE[, 2]),
            "ensembl_id"
          ]
          TERM2GENE <- unnest_fun(
            TERM2GENE,
            cols = "ensembl_id-converted",
            keep_empty = FALSE
          )
          TERM2GENE <- TERM2GENE[, c("Term", "ensembl_id-converted")]
          colnames(TERM2GENE) <- c("Term", "ensembl_id")
          TERM2NAME <- TERM2NAME[
            TERM2NAME[["Term"]] %in% TERM2GENE[["Term"]], ,
            drop = FALSE
          ]
        }

        db_info[["TERM2GENE"]] <- unique(TERM2GENE)
        db_info[["TERM2NAME"]] <- unique(TERM2NAME)
        version <- paste0(
          db_info[["version"]],
          "(converted from ",
          sp_from,
          ")"
        )
        db_info[["version"]] <- version
        db_list[[sps]][[term]] <- db_info
        default_id_types[[term]] <- "ensembl_id"
        # save cache
        R.cache::saveCache(
          db_list[[sps]][[term]],
          key = list(version, sps, term),
          comment = paste0(
            version,
            " nterm:",
            length(TERM2NAME[[1]]),
            "|",
            sps,
            "-",
            term
          )
        )
      }
    }

    # Convert ID types
    for (term in names(db_list[[sps]])) {
      IDtypes <- db_IDtypes[
        !db_IDtypes %in% colnames(db_list[[sps]][[term]][["TERM2GENE"]])
      ]
      if (length(IDtypes) > 0) {
        log_message("Convert ID types for the database: ", term)
        TERM2GENE <- db_list[[sps]][[term]][["TERM2GENE"]]
        TERM2NAME <- db_list[[sps]][[term]][["TERM2NAME"]]
        if (is.na(default_id_types[term])) {
          IDtype <- colnames(TERM2GENE)[2]
        } else {
          IDtype <- default_id_types[[term]]
        }
        if (grepl("MSigDB_", term)) {
          map <- db_list[[sps]][["MSigDB"]][["TERM2GENE"]][, -1, drop = FALSE]
          map <- stats::aggregate(
            map,
            by = list(map[[1]]),
            FUN = function(x) list(unique(x))
          )
          rownames(map) <- map[, 1]
        } else {
          res <- GeneConvert(
            geneID = as.character(unique(TERM2GENE[, 2])),
            geneID_from_IDtype = IDtype,
            geneID_to_IDtype = IDtypes,
            species_from = sps,
            species_to = sps,
            Ensembl_version = Ensembl_version,
            mirror = mirror,
            biomart = biomart,
            max_tries = max_tries
          )
          if (is.null(res$geneID_res)) {
            log_message(
              "Failed to convert ID types for the database: ",
              term,
              message_type = "warning"
            )
            next
          }
          map <- res$geneID_collapse
        }
        for (type in IDtypes) {
          TERM2GENE[[type]] <- map[as.character(TERM2GENE[, 2]), type]
          TERM2GENE <- unnest_fun(TERM2GENE, cols = type, keep_empty = TRUE)
        }
        db_list[[sps]][[term]][["TERM2GENE"]] <- TERM2GENE
        # save cache
        version <- db_list[[sps]][[term]][["version"]]
        R.cache::saveCache(
          db_list[[sps]][[term]],
          key = list(version, sps, term),
          comment = paste0(
            version,
            " nterm:",
            length(TERM2NAME[[1]]),
            "|",
            sps,
            "-",
            term
          )
        )
      }
    }
  }
  return(db_list)
}

#' @title List cached databases
#'
#' @description
#' Retrieves information about databases based on a given species and database name.
#'
#' @md
#' @param species The species for which to retrieve database information.
#' Default is `"Homo_sapiens"`.
#' @param db The pattern to match against the database names.
#' Default is `NULL`, which matches all databases.
#'
#' @return A data frame containing information about the databases.
#'
#' @seealso [PrepareDB]
#'
#' @export
#' @examples
#' ListDB(species = "Homo_sapiens")
#' ListDB(species = "Mus_musculus", db = "GO_BP")
ListDB <- function(
    species = "Homo_sapiens",
    db = NULL) {
  stopifnot(length(species) == 1)
  pathnames <- dir(
    path = R.cache::getCacheRootPath(),
    pattern = "[.]Rcache$",
    full.names = TRUE
  )
  if (length(pathnames) == 0) {
    return(NULL)
  }
  dbinfo <- lapply(
    pathnames, function(x) {
      info <- R.cache::readCacheHeader(x)
      info[["date"]] <- as.character(info[["timestamp"]])
      info[["db_version"]] <- strsplit(info[["comment"]], "\\|")[[1]][1]
      info[["db_name"]] <- strsplit(info[["comment"]], "\\|")[[1]][2]
      info
    }
  )
  dbinfo <- do.call(rbind.data.frame, dbinfo)
  dbinfo[["file"]] <- pathnames

  if (is.null(db)) {
    db <- ".*"
  }
  patterns <- paste0("^", species, "-", db, "$")
  dbinfo <- dbinfo[
    unlist(lapply(patterns, function(pat) grep(pat, dbinfo[["db_name"]]))), ,
    drop = FALSE
  ]
  dbinfo <- dbinfo[
    order(dbinfo[["timestamp"]], decreasing = TRUE), ,
    drop = FALSE
  ]
  rownames(dbinfo) <- NULL
  return(dbinfo)
}
